{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# T1D Data Process with Low Complexity Filter\n",
    "Starting with previously trimmed fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pymongo\n",
    "from glob import glob\n",
    "import gzip\n",
    "import re\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import pandas as pd\n",
    "from joblib import Parallel, delayed\n",
    "from multiprocessing import cpu_count\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S01.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S02.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S03.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S04.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S05.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S06.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S07.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S08.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S09.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S10.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S11.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S12.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S13.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S14.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S15.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S16.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S17.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S18.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S19.trim.fastq.gz\n",
      "fqtrim -l 15 -D -p 12 -o dust.fastq.gz S20.trim.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for f in *.trim.fastq.gz; do\n",
    "out=\"dust.fastq.gz\"\n",
    "echo \"fqtrim -l 15 -D -p 12 -o $out $f\"\n",
    "fqtrim -l 15 -D -p 12 -o $out $f\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S01.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S01.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S02.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S02.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S03.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S03.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S04.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S04.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S05.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S05.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S06.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S06.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S07.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S07.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S08.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S08.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S09.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S09.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S10.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S10.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S11.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S11.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S12.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S12.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S13.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S13.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S14.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S14.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S15.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S15.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S16.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S16.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S17.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S17.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S18.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S18.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S19.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S19.bam\n",
      "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U S20.trim.dust.fastq.gz | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o S20.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1015469 reads; of these:\n",
      "  1015469 (100.00%) were unpaired; of these:\n",
      "    229985 (22.65%) aligned 0 times\n",
      "    236877 (23.33%) aligned exactly 1 time\n",
      "    548607 (54.02%) aligned >1 times\n",
      "77.35% overall alignment rate\n",
      "3269311 reads; of these:\n",
      "  3269311 (100.00%) were unpaired; of these:\n",
      "    740256 (22.64%) aligned 0 times\n",
      "    630427 (19.28%) aligned exactly 1 time\n",
      "    1898628 (58.07%) aligned >1 times\n",
      "77.36% overall alignment rate\n",
      "2422437 reads; of these:\n",
      "  2422437 (100.00%) were unpaired; of these:\n",
      "    1121984 (46.32%) aligned 0 times\n",
      "    225647 (9.31%) aligned exactly 1 time\n",
      "    1074806 (44.37%) aligned >1 times\n",
      "53.68% overall alignment rate\n",
      "2667977 reads; of these:\n",
      "  2667977 (100.00%) were unpaired; of these:\n",
      "    1543527 (57.85%) aligned 0 times\n",
      "    139284 (5.22%) aligned exactly 1 time\n",
      "    985166 (36.93%) aligned >1 times\n",
      "42.15% overall alignment rate\n",
      "1872931 reads; of these:\n",
      "  1872931 (100.00%) were unpaired; of these:\n",
      "    744989 (39.78%) aligned 0 times\n",
      "    151867 (8.11%) aligned exactly 1 time\n",
      "    976075 (52.11%) aligned >1 times\n",
      "60.22% overall alignment rate\n",
      "2566936 reads; of these:\n",
      "  2566936 (100.00%) were unpaired; of these:\n",
      "    2335992 (91.00%) aligned 0 times\n",
      "    27874 (1.09%) aligned exactly 1 time\n",
      "    203070 (7.91%) aligned >1 times\n",
      "9.00% overall alignment rate\n",
      "3361981 reads; of these:\n",
      "  3361981 (100.00%) were unpaired; of these:\n",
      "    549570 (16.35%) aligned 0 times\n",
      "    718167 (21.36%) aligned exactly 1 time\n",
      "    2094244 (62.29%) aligned >1 times\n",
      "83.65% overall alignment rate\n",
      "4707719 reads; of these:\n",
      "  4707719 (100.00%) were unpaired; of these:\n",
      "    638901 (13.57%) aligned 0 times\n",
      "    1313465 (27.90%) aligned exactly 1 time\n",
      "    2755353 (58.53%) aligned >1 times\n",
      "86.43% overall alignment rate\n",
      "2117071 reads; of these:\n",
      "  2117071 (100.00%) were unpaired; of these:\n",
      "    723019 (34.15%) aligned 0 times\n",
      "    315541 (14.90%) aligned exactly 1 time\n",
      "    1078511 (50.94%) aligned >1 times\n",
      "65.85% overall alignment rate\n",
      "1531589 reads; of these:\n",
      "  1531589 (100.00%) were unpaired; of these:\n",
      "    475627 (31.05%) aligned 0 times\n",
      "    215767 (14.09%) aligned exactly 1 time\n",
      "    840195 (54.86%) aligned >1 times\n",
      "68.95% overall alignment rate\n",
      "2130681 reads; of these:\n",
      "  2130681 (100.00%) were unpaired; of these:\n",
      "    642653 (30.16%) aligned 0 times\n",
      "    347536 (16.31%) aligned exactly 1 time\n",
      "    1140492 (53.53%) aligned >1 times\n",
      "69.84% overall alignment rate\n",
      "3235562 reads; of these:\n",
      "  3235562 (100.00%) were unpaired; of these:\n",
      "    491036 (15.18%) aligned 0 times\n",
      "    639921 (19.78%) aligned exactly 1 time\n",
      "    2104605 (65.05%) aligned >1 times\n",
      "84.82% overall alignment rate\n",
      "982151 reads; of these:\n",
      "  982151 (100.00%) were unpaired; of these:\n",
      "    271890 (27.68%) aligned 0 times\n",
      "    125146 (12.74%) aligned exactly 1 time\n",
      "    585115 (59.57%) aligned >1 times\n",
      "72.32% overall alignment rate\n",
      "2954328 reads; of these:\n",
      "  2954328 (100.00%) were unpaired; of these:\n",
      "    452402 (15.31%) aligned 0 times\n",
      "    426904 (14.45%) aligned exactly 1 time\n",
      "    2075022 (70.24%) aligned >1 times\n",
      "84.69% overall alignment rate\n",
      "22929 reads; of these:\n",
      "  22929 (100.00%) were unpaired; of these:\n",
      "    11525 (50.26%) aligned 0 times\n",
      "    1736 (7.57%) aligned exactly 1 time\n",
      "    9668 (42.16%) aligned >1 times\n",
      "49.74% overall alignment rate\n",
      "2689776 reads; of these:\n",
      "  2689776 (100.00%) were unpaired; of these:\n",
      "    367831 (13.68%) aligned 0 times\n",
      "    554118 (20.60%) aligned exactly 1 time\n",
      "    1767827 (65.72%) aligned >1 times\n",
      "86.32% overall alignment rate\n",
      "948554 reads; of these:\n",
      "  948554 (100.00%) were unpaired; of these:\n",
      "    157247 (16.58%) aligned 0 times\n",
      "    147876 (15.59%) aligned exactly 1 time\n",
      "    643431 (67.83%) aligned >1 times\n",
      "83.42% overall alignment rate\n",
      "3781304 reads; of these:\n",
      "  3781304 (100.00%) were unpaired; of these:\n",
      "    1202797 (31.81%) aligned 0 times\n",
      "    625780 (16.55%) aligned exactly 1 time\n",
      "    1952727 (51.64%) aligned >1 times\n",
      "68.19% overall alignment rate\n",
      "2164708 reads; of these:\n",
      "  2164708 (100.00%) were unpaired; of these:\n",
      "    573333 (26.49%) aligned 0 times\n",
      "    214416 (9.91%) aligned exactly 1 time\n",
      "    1376959 (63.61%) aligned >1 times\n",
      "73.51% overall alignment rate\n",
      "3903034 reads; of these:\n",
      "  3903034 (100.00%) were unpaired; of these:\n",
      "    1083300 (27.76%) aligned 0 times\n",
      "    579130 (14.84%) aligned exactly 1 time\n",
      "    2240604 (57.41%) aligned >1 times\n",
      "72.24% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for f in *.trim.dust.fastq.gz; do\n",
    "out=${f/.trim.dust.fastq.gz/.bam}\n",
    "echo \"bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U $f | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o $out\";\n",
    "bowtie2 --sensitive --end-to-end -N 1 -p 12 -x /avicenna/genomes/hg38/hg38 -U $f | samtools view - -Sb  -h -t /avicenna/genomes/hg38/hg38.fa.fai -o $out ;\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools sort -o S01.srt.bam S01.bam\n",
      "samtools sort -o S02.srt.bam S02.bam\n",
      "samtools sort -o S03.srt.bam S03.bam\n",
      "samtools sort -o S04.srt.bam S04.bam\n",
      "samtools sort -o S05.srt.bam S05.bam\n",
      "samtools sort -o S06.srt.bam S06.bam\n",
      "samtools sort -o S07.srt.bam S07.bam\n",
      "samtools sort -o S08.srt.bam S08.bam\n",
      "samtools sort -o S09.srt.bam S09.bam\n",
      "samtools sort -o S10.srt.bam S10.bam\n",
      "samtools sort -o S11.srt.bam S11.bam\n",
      "samtools sort -o S12.srt.bam S12.bam\n",
      "samtools sort -o S13.srt.bam S13.bam\n",
      "samtools sort -o S14.srt.bam S14.bam\n",
      "samtools sort -o S15.srt.bam S15.bam\n",
      "samtools sort -o S16.srt.bam S16.bam\n",
      "samtools sort -o S17.srt.bam S17.bam\n",
      "samtools sort -o S18.srt.bam S18.bam\n",
      "samtools sort -o S19.srt.bam S19.bam\n",
      "samtools sort -o S20.srt.bam S20.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for f in *.bam; do\n",
    "out=${f/.bam/.srt.bam};\n",
    "echo \"samtools sort -o $out $f\";\n",
    "samtools sort -@ 4 -m 2G -o $out $f\n",
    "samtools index $out;\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "umi_tools dedup -I S01.srt.bam -S S01.srt.dd.bam\n",
      "umi_tools dedup -I S02.srt.bam -S S02.srt.dd.bam\n",
      "umi_tools dedup -I S03.srt.bam -S S03.srt.dd.bam\n",
      "umi_tools dedup -I S04.srt.bam -S S04.srt.dd.bam\n",
      "umi_tools dedup -I S05.srt.bam -S S05.srt.dd.bam\n",
      "umi_tools dedup -I S06.srt.bam -S S06.srt.dd.bam\n",
      "umi_tools dedup -I S07.srt.bam -S S07.srt.dd.bam\n",
      "umi_tools dedup -I S08.srt.bam -S S08.srt.dd.bam\n",
      "umi_tools dedup -I S09.srt.bam -S S09.srt.dd.bam\n",
      "umi_tools dedup -I S10.srt.bam -S S10.srt.dd.bam\n",
      "umi_tools dedup -I S11.srt.bam -S S11.srt.dd.bam\n",
      "umi_tools dedup -I S12.srt.bam -S S12.srt.dd.bam\n",
      "umi_tools dedup -I S13.srt.bam -S S13.srt.dd.bam\n",
      "umi_tools dedup -I S14.srt.bam -S S14.srt.dd.bam\n",
      "umi_tools dedup -I S15.srt.bam -S S15.srt.dd.bam\n",
      "umi_tools dedup -I S16.srt.bam -S S16.srt.dd.bam\n",
      "umi_tools dedup -I S17.srt.bam -S S17.srt.dd.bam\n",
      "umi_tools dedup -I S18.srt.bam -S S18.srt.dd.bam\n",
      "umi_tools dedup -I S19.srt.bam -S S19.srt.dd.bam\n",
      "umi_tools dedup -I S20.srt.bam -S S20.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S18.srt.bam -S S18.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- cfec06b7-f88b-4c68-ad76-13cb79b08183\n",
      "# pid: 3325, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S18.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S18.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,821 INFO command: dedup -I S18.srt.bam -S S18.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S20.srt.bam -S S20.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 4bb1dee8-f8d7-4645-918b-400d1971d433\n",
      "# pid: 3327, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S20.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S20.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,840 INFO command: dedup -I S20.srt.bam -S S20.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S12.srt.bam -S S12.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- c6b57d14-edee-4193-b996-4350a56b4444\n",
      "# pid: 3319, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S12.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S12.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,884 INFO command: dedup -I S12.srt.bam -S S12.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S04.srt.bam -S S04.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 0612eb08-9c68-48c9-aae4-c2b6104ab138\n",
      "# pid: 3311, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S04.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S04.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,894 INFO command: dedup -I S04.srt.bam -S S04.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S07.srt.bam -S S07.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 6f7b3750-a7c7-4ade-8dac-266b1ce46bc8\n",
      "# pid: 3314, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S07.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S07.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,919 INFO command: dedup -I S07.srt.bam -S S07.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S10.srt.bam -S S10.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 62602f78-9b17-4fc7-a2f0-138219a71f24\n",
      "# pid: 3317, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S10.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S10.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,926 INFO command: dedup -I S10.srt.bam -S S10.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S05.srt.bam -S S05.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 40248c6b-6665-4678-9fb8-8e48017511b5\n",
      "# pid: 3312, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S05.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S05.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,937 INFO command: dedup -I S05.srt.bam -S S05.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S14.srt.bam -S S14.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 8fac16bd-f125-448c-b1ea-21f3f3cd3c8d\n",
      "# pid: 3321, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S14.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S14.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,946 INFO command: dedup -I S14.srt.bam -S S14.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S13.srt.bam -S S13.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- bb5bc5a5-ce07-4138-954c-2e8280fcf249\n",
      "# pid: 3320, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S13.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S13.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,954 INFO command: dedup -I S13.srt.bam -S S13.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S17.srt.bam -S S17.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- f318f2d2-e3a6-4a72-a9e9-2df99c3f10a8\n",
      "# pid: 3324, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S17.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S17.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,973 INFO command: dedup -I S17.srt.bam -S S17.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S15.srt.bam -S S15.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 859cb821-5add-4f08-b568-08d0b7bbfda3\n",
      "# pid: 3322, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S15.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S15.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,976 INFO command: dedup -I S15.srt.bam -S S15.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S09.srt.bam -S S09.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:01 2020 on bch-svr.ucsf.edu -- 06311c31-2a03-4f3b-b52f-f28abe088ba5\n",
      "# pid: 3316, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S09.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S09.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:01,981 INFO command: dedup -I S09.srt.bam -S S09.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S01.srt.bam -S S01.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- 4dde2f71-4463-4e77-8ac9-d719e468d215\n",
      "# pid: 3308, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S01.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S01.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,001 INFO command: dedup -I S01.srt.bam -S S01.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S11.srt.bam -S S11.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- 64536224-147f-4ea5-80a8-98943aaf8661\n",
      "# pid: 3318, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S11.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S11.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,007 INFO command: dedup -I S11.srt.bam -S S11.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S06.srt.bam -S S06.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- cff9393d-9b0e-4533-a40e-88f723455e03\n",
      "# pid: 3313, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S06.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S06.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,009 INFO command: dedup -I S06.srt.bam -S S06.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S16.srt.bam -S S16.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- 27130992-ad40-4912-a014-5eddabccbfd2\n",
      "# pid: 3323, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S16.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S16.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S08.srt.bam -S S08.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- fb1fe9c2-6852-43bb-9ee0-b00759c696dd\n",
      "# pid: 3315, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S08.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S08.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,013 INFO command: dedup -I S16.srt.bam -S S16.srt.dd.bam\n",
      "2020-04-09 11:09:02,013 INFO command: dedup -I S08.srt.bam -S S08.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S03.srt.bam -S S03.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- 8f53ee7d-e0d8-4849-8531-7b387758b9fb\n",
      "# pid: 3310, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S03.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S03.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,035 INFO command: dedup -I S03.srt.bam -S S03.srt.dd.bam\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S02.srt.bam -S S02.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- 8ff07e19-7bde-4eba-b3b5-e81483c6a179\n",
      "# pid: 3309, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S02.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S02.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "# UMI-tools version: 1.0.1\n",
      "# output generated by dedup -I S19.srt.bam -S S19.srt.dd.bam\n",
      "# job started at Thu Apr  9 11:09:02 2020 on bch-svr.ucsf.edu -- c6cce8d1-9524-45ae-b695-da061f47582b\n",
      "# pid: 3326, system: Linux 4.18.0-21-generic #22-Ubuntu SMP Wed May 15 13:13:21 UTC 2019 x86_64\n",
      "# assigned_tag                            : None\n",
      "# cell_tag                                : None\n",
      "# cell_tag_delim                          : None\n",
      "# cell_tag_split                          : -\n",
      "# chimeric_pairs                          : use\n",
      "# chrom                                   : None\n",
      "# compresslevel                           : 6\n",
      "# detection_method                        : None\n",
      "# gene_tag                                : None\n",
      "# gene_transcript_map                     : None\n",
      "# get_umi_method                          : read_id\n",
      "# ignore_umi                              : False\n",
      "# in_sam                                  : False\n",
      "# log2stderr                              : False\n",
      "# loglevel                                : 1\n",
      "# mapping_quality                         : 0\n",
      "# method                                  : directional\n",
      "# no_sort_output                          : False\n",
      "# out_sam                                 : False\n",
      "# output_unmapped                         : False\n",
      "# paired                                  : False\n",
      "# per_cell                                : False\n",
      "# per_contig                              : False\n",
      "# per_gene                                : False\n",
      "# random_seed                             : None\n",
      "# read_length                             : False\n",
      "# short_help                              : None\n",
      "# skip_regex                              : ^(__|Unassigned)\n",
      "# soft_clip_threshold                     : 4\n",
      "# spliced                                 : False\n",
      "# stats                                   : False\n",
      "# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>\n",
      "# stdin                                   : <_io.TextIOWrapper name='S19.srt.bam' mode='r' encoding='UTF-8'>\n",
      "# stdlog                                  : <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>\n",
      "# stdout                                  : <_io.TextIOWrapper name='S19.srt.dd.bam' mode='w' encoding='UTF-8'>\n",
      "# subset                                  : None\n",
      "# threshold                               : 1\n",
      "# timeit_file                             : None\n",
      "# timeit_header                           : None\n",
      "# timeit_name                             : all\n",
      "# tmpdir                                  : None\n",
      "# umi_sep                                 : _\n",
      "# umi_tag                                 : RX\n",
      "# umi_tag_delim                           : None\n",
      "# umi_tag_split                           : None\n",
      "# unmapped_reads                          : discard\n",
      "# unpaired_reads                          : use\n",
      "# whole_contig                            : False\n",
      "2020-04-09 11:09:02,054 INFO command: dedup -I S02.srt.bam -S S02.srt.dd.bam\n",
      "2020-04-09 11:09:02,054 INFO command: dedup -I S19.srt.bam -S S19.srt.dd.bam\n",
      "2020-04-09 11:09:02,624 INFO Reads: Input Reads: 11404\n",
      "2020-04-09 11:09:02,624 INFO Number of reads out: 10178\n",
      "2020-04-09 11:09:02,624 INFO Total number of positions deduplicated: 4415\n",
      "2020-04-09 11:09:02,625 INFO Mean number of unique UMIs per position: 2.53\n",
      "2020-04-09 11:09:02,625 INFO Max. number of unique UMIs per position: 907\n",
      "# job finished in 0 seconds at Thu Apr  9 11:09:02 2020 --  1.25  0.22  0.00  0.00 -- 859cb821-5add-4f08-b568-08d0b7bbfda3\n",
      "2020-04-09 11:09:05,096 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:05,593 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:06,253 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:07,380 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:08,356 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:08,635 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:08,745 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:08,772 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:09,248 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:09,373 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:09,962 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:10,209 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:11,018 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:11,604 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:12,112 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:12,383 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:12,465 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:13,507 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:13,914 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:14,159 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:15,236 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:15,607 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:17,849 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:19,028 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:19,291 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:19,411 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:19,741 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:19,806 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:20,141 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:20,657 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:20,911 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:20,940 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:21,933 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:23,056 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:24,093 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:24,167 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:24,559 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:25,054 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:25,134 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:26,315 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:26,438 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:26,859 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:28,290 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:28,542 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:28,676 INFO Reads: Input Reads: 230944\n",
      "2020-04-09 11:09:28,676 INFO Number of reads out: 171699\n",
      "2020-04-09 11:09:28,676 INFO Total number of positions deduplicated: 46593\n",
      "2020-04-09 11:09:28,676 INFO Mean number of unique UMIs per position: 4.81\n",
      "2020-04-09 11:09:28,676 INFO Max. number of unique UMIs per position: 6514\n",
      "# job finished in 26 seconds at Thu Apr  9 11:09:28 2020 -- 26.90  0.54  0.00  0.00 -- cff9393d-9b0e-4533-a40e-88f723455e03\n",
      "2020-04-09 11:09:29,088 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:29,736 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:29,864 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:30,120 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:31,157 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:32,102 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:32,766 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:34,061 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:34,257 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:34,669 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:34,741 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:35,335 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:36,519 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:36,855 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:37,913 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:37,965 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:38,057 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:38,461 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:40,147 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:40,292 INFO Written out 800000 reads\n",
      "2020-04-09 11:09:40,492 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:40,800 INFO Written out 200000 reads\n",
      "2020-04-09 11:09:41,129 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:41,574 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:42,464 INFO Reads: Input Reads: 785484\n",
      "2020-04-09 11:09:42,464 INFO Number of reads out: 736451\n",
      "2020-04-09 11:09:42,464 INFO Total number of positions deduplicated: 443760\n",
      "2020-04-09 11:09:42,464 INFO Mean number of unique UMIs per position: 1.76\n",
      "2020-04-09 11:09:42,465 INFO Max. number of unique UMIs per position: 2545\n",
      "# job finished in 40 seconds at Thu Apr  9 11:09:42 2020 -- 39.83  1.11  0.00  0.00 -- 4dde2f71-4463-4e77-8ac9-d719e468d215\n",
      "2020-04-09 11:09:43,049 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:43,560 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:43,616 INFO Written out 100000 reads\n",
      "2020-04-09 11:09:44,453 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:45,379 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:45,713 INFO Written out 300000 reads\n",
      "2020-04-09 11:09:45,991 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:46,327 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:47,022 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:47,325 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:47,953 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:48,090 INFO Written out 900000 reads\n",
      "2020-04-09 11:09:48,228 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:48,941 INFO Written out 800000 reads\n",
      "2020-04-09 11:09:49,201 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:49,801 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:50,375 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:09:50,887 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:51,470 INFO Reads: Input Reads: 710261\n",
      "2020-04-09 11:09:51,470 INFO Number of reads out: 615600\n",
      "2020-04-09 11:09:51,470 INFO Total number of positions deduplicated: 311352\n",
      "2020-04-09 11:09:51,470 INFO Mean number of unique UMIs per position: 2.25\n",
      "2020-04-09 11:09:51,470 INFO Max. number of unique UMIs per position: 4505\n",
      "# job finished in 49 seconds at Thu Apr  9 11:09:51 2020 -- 49.22  1.00  0.00  0.00 -- bb5bc5a5-ce07-4138-954c-2e8280fcf249\n",
      "2020-04-09 11:09:52,543 INFO Written out 400000 reads\n",
      "2020-04-09 11:09:53,296 INFO Written out 800000 reads\n",
      "2020-04-09 11:09:53,371 INFO Written out 600000 reads\n",
      "2020-04-09 11:09:54,952 INFO Written out 1000000 reads\n",
      "2020-04-09 11:09:55,220 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:09:55,722 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:56,051 INFO Written out 800000 reads\n",
      "2020-04-09 11:09:56,232 INFO Written out 900000 reads\n",
      "2020-04-09 11:09:56,281 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:56,926 INFO Written out 700000 reads\n",
      "2020-04-09 11:09:58,112 INFO Written out 800000 reads\n",
      "2020-04-09 11:09:58,970 INFO Written out 500000 reads\n",
      "2020-04-09 11:09:59,433 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:09:59,552 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:01,014 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:01,113 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:01,619 INFO Written out 300000 reads\n",
      "2020-04-09 11:10:02,843 INFO Written out 200000 reads\n",
      "2020-04-09 11:10:03,036 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:04,104 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:05,607 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:05,703 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:07,095 INFO Written out 400000 reads\n",
      "2020-04-09 11:10:07,097 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:08,507 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:08,545 INFO Written out 600000 reads\n",
      "2020-04-09 11:10:09,564 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:09,630 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:09,865 INFO Written out 300000 reads\n",
      "2020-04-09 11:10:10,087 INFO Written out 500000 reads\n",
      "2020-04-09 11:10:11,270 INFO Written out 200000 reads\n",
      "2020-04-09 11:10:11,288 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:12,345 INFO Written out 200000 reads\n",
      "2020-04-09 11:10:13,035 INFO Written out 300000 reads\n",
      "2020-04-09 11:10:13,623 INFO Written out 600000 reads\n",
      "2020-04-09 11:10:14,118 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:14,123 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:14,472 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:15,423 INFO Written out 700000 reads\n",
      "2020-04-09 11:10:16,540 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:16,916 INFO Written out 1300000 reads\n",
      "2020-04-09 11:10:19,912 INFO Reads: Input Reads: 1127942\n",
      "2020-04-09 11:10:19,912 INFO Number of reads out: 884096\n",
      "2020-04-09 11:10:19,912 INFO Total number of positions deduplicated: 318105\n",
      "2020-04-09 11:10:19,912 INFO Mean number of unique UMIs per position: 3.46\n",
      "2020-04-09 11:10:19,912 INFO Max. number of unique UMIs per position: 5579\n",
      "# job finished in 77 seconds at Thu Apr  9 11:10:19 2020 -- 77.40  1.27  0.00  0.00 -- 40248c6b-6665-4678-9fb8-8e48017511b5\n",
      "2020-04-09 11:10:20,060 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:20,713 INFO Written out 700000 reads\n",
      "2020-04-09 11:10:20,846 INFO Written out 1400000 reads\n",
      "2020-04-09 11:10:20,914 INFO Written out 400000 reads\n",
      "2020-04-09 11:10:22,037 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:22,260 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:22,836 INFO Written out 400000 reads\n",
      "2020-04-09 11:10:23,309 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:25,547 INFO Written out 500000 reads\n",
      "2020-04-09 11:10:27,724 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:27,737 INFO Written out 1300000 reads\n",
      "2020-04-09 11:10:28,494 INFO Written out 1300000 reads\n",
      "2020-04-09 11:10:28,977 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:30,171 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:30,747 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:30,814 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:31,107 INFO Written out 500000 reads\n",
      "2020-04-09 11:10:31,993 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:32,371 INFO Written out 600000 reads\n",
      "2020-04-09 11:10:33,484 INFO Written out 1400000 reads\n",
      "2020-04-09 11:10:34,569 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:34,671 INFO Written out 1400000 reads\n",
      "2020-04-09 11:10:34,681 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:35,430 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:35,736 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:35,829 INFO Written out 600000 reads\n",
      "2020-04-09 11:10:38,139 INFO Written out 1500000 reads\n",
      "2020-04-09 11:10:38,452 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:38,761 INFO Written out 900000 reads\n",
      "2020-04-09 11:10:38,914 INFO Written out 700000 reads\n",
      "2020-04-09 11:10:39,566 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:40,106 INFO Written out 1500000 reads\n",
      "2020-04-09 11:10:41,501 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:41,950 INFO Written out 1600000 reads\n",
      "2020-04-09 11:10:42,391 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:42,510 INFO Written out 1000000 reads\n",
      "2020-04-09 11:10:42,792 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:42,855 INFO Written out 300000 reads\n",
      "2020-04-09 11:10:43,728 INFO Written out 700000 reads\n",
      "2020-04-09 11:10:45,898 INFO Reads: Input Reads: 1394052\n",
      "2020-04-09 11:10:45,899 INFO Number of reads out: 1194361\n",
      "2020-04-09 11:10:45,899 INFO Total number of positions deduplicated: 626310\n",
      "2020-04-09 11:10:45,899 INFO Mean number of unique UMIs per position: 2.19\n",
      "2020-04-09 11:10:45,899 INFO Max. number of unique UMIs per position: 11368\n",
      "# job finished in 103 seconds at Thu Apr  9 11:10:45 2020 -- 102.86  1.86  0.00  0.00 -- 06311c31-2a03-4f3b-b52f-f28abe088ba5\n",
      "2020-04-09 11:10:47,772 INFO Written out 1500000 reads\n",
      "2020-04-09 11:10:47,785 INFO Reads: Input Reads: 1300453\n",
      "2020-04-09 11:10:47,785 INFO Number of reads out: 1029484\n",
      "2020-04-09 11:10:47,785 INFO Total number of positions deduplicated: 394681\n",
      "2020-04-09 11:10:47,785 INFO Mean number of unique UMIs per position: 3.22\n",
      "2020-04-09 11:10:47,785 INFO Max. number of unique UMIs per position: 10759\n",
      "# job finished in 105 seconds at Thu Apr  9 11:10:47 2020 -- 104.62  1.91  0.00  0.00 -- 8f53ee7d-e0d8-4849-8531-7b387758b9fb\n",
      "2020-04-09 11:10:49,258 INFO Written out 700000 reads\n",
      "2020-04-09 11:10:49,408 INFO Written out 400000 reads\n",
      "2020-04-09 11:10:49,620 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:50,386 INFO Written out 1600000 reads\n",
      "2020-04-09 11:10:50,528 INFO Written out 1700000 reads\n",
      "2020-04-09 11:10:50,862 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:51,196 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:52,036 INFO Written out 1100000 reads\n",
      "2020-04-09 11:10:54,804 INFO Written out 800000 reads\n",
      "2020-04-09 11:10:55,557 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:10:55,583 INFO Written out 1800000 reads\n",
      "2020-04-09 11:10:55,788 INFO Written out 1700000 reads\n",
      "2020-04-09 11:10:56,279 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:56,463 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:10:58,863 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:59,095 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:59,279 INFO Written out 1600000 reads\n",
      "2020-04-09 11:10:59,331 INFO Written out 1200000 reads\n",
      "2020-04-09 11:10:59,462 INFO Written out 1300000 reads\n",
      "2020-04-09 11:11:03,990 INFO Reads: Input Reads: 1055962\n",
      "2020-04-09 11:11:03,991 INFO Number of reads out: 858287\n",
      "2020-04-09 11:11:03,991 INFO Total number of positions deduplicated: 417317\n",
      "2020-04-09 11:11:03,991 INFO Mean number of unique UMIs per position: 2.46\n",
      "2020-04-09 11:11:03,991 INFO Max. number of unique UMIs per position: 15613\n",
      "# job finished in 122 seconds at Thu Apr  9 11:11:03 2020 -- 120.68  1.93  0.00  0.00 -- 62602f78-9b17-4fc7-a2f0-138219a71f24\n",
      "2020-04-09 11:11:05,398 INFO Written out 1300000 reads\n",
      "2020-04-09 11:11:05,464 INFO Reads: Input Reads: 1488028\n",
      "2020-04-09 11:11:05,465 INFO Number of reads out: 1225931\n",
      "2020-04-09 11:11:05,465 INFO Total number of positions deduplicated: 599090\n",
      "2020-04-09 11:11:05,465 INFO Mean number of unique UMIs per position: 2.42\n",
      "2020-04-09 11:11:05,465 INFO Max. number of unique UMIs per position: 8936\n",
      "# job finished in 123 seconds at Thu Apr  9 11:11:05 2020 -- 121.92  2.21  0.00  0.00 -- 64536224-147f-4ea5-80a8-98943aaf8661\n",
      "2020-04-09 11:11:05,508 INFO Written out 500000 reads\n",
      "2020-04-09 11:11:05,725 INFO Written out 1800000 reads\n",
      "2020-04-09 11:11:05,931 INFO Written out 800000 reads\n",
      "2020-04-09 11:11:06,303 INFO Written out 1400000 reads\n",
      "2020-04-09 11:11:07,025 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:11:07,242 INFO Written out 1300000 reads\n",
      "2020-04-09 11:11:07,725 INFO Written out 1900000 reads\n",
      "2020-04-09 11:11:09,379 INFO Written out 1500000 reads\n",
      "2020-04-09 11:11:11,937 INFO Written out 1700000 reads\n",
      "2020-04-09 11:11:12,266 INFO Written out 600000 reads\n",
      "2020-04-09 11:11:13,511 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:11:13,912 INFO Written out 1900000 reads\n",
      "2020-04-09 11:11:14,889 INFO Written out 1400000 reads\n",
      "2020-04-09 11:11:15,075 INFO Written out 2000000 reads\n",
      "2020-04-09 11:11:15,105 INFO Written out 1400000 reads\n",
      "2020-04-09 11:11:17,144 INFO Written out 2000000 reads\n",
      "2020-04-09 11:11:20,249 INFO Written out 1800000 reads\n",
      "2020-04-09 11:11:22,172 INFO Written out 2100000 reads\n",
      "2020-04-09 11:11:22,997 INFO Written out 1500000 reads\n",
      "2020-04-09 11:11:23,488 INFO Written out 2100000 reads\n",
      "2020-04-09 11:11:23,617 INFO Written out 1600000 reads\n",
      "2020-04-09 11:11:23,625 INFO Written out 2200000 reads\n",
      "2020-04-09 11:11:23,749 INFO Written out 900000 reads\n",
      "2020-04-09 11:11:25,821 INFO Written out 1500000 reads\n",
      "2020-04-09 11:11:26,277 INFO Written out 900000 reads\n",
      "2020-04-09 11:11:26,679 INFO Written out 2300000 reads\n",
      "2020-04-09 11:11:29,721 INFO Written out 2200000 reads\n",
      "2020-04-09 11:11:29,860 INFO Written out 2400000 reads\n",
      "2020-04-09 11:11:30,030 INFO Written out 1000000 reads\n",
      "2020-04-09 11:11:30,923 INFO Written out 1600000 reads\n",
      "2020-04-09 11:11:32,067 INFO Written out 700000 reads\n",
      "2020-04-09 11:11:32,948 INFO Written out 2500000 reads\n",
      "2020-04-09 11:11:34,233 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:11:34,625 INFO Written out 1000000 reads\n",
      "2020-04-09 11:11:35,568 INFO Written out 1700000 reads\n",
      "2020-04-09 11:11:36,411 INFO Written out 1600000 reads\n",
      "2020-04-09 11:11:36,585 INFO Written out 1900000 reads\n",
      "2020-04-09 11:11:36,705 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:11:36,968 INFO Written out 2600000 reads\n",
      "2020-04-09 11:11:37,850 INFO Written out 800000 reads\n",
      "2020-04-09 11:11:38,346 INFO Reads: Input Reads: 2529055\n",
      "2020-04-09 11:11:38,346 INFO Number of reads out: 2203144\n",
      "2020-04-09 11:11:38,346 INFO Total number of positions deduplicated: 1242697\n",
      "2020-04-09 11:11:38,346 INFO Mean number of unique UMIs per position: 2.01\n",
      "2020-04-09 11:11:38,347 INFO Max. number of unique UMIs per position: 10623\n",
      "# job finished in 156 seconds at Thu Apr  9 11:11:38 2020 -- 154.03  3.20  0.00  0.00 -- 8ff07e19-7bde-4eba-b3b5-e81483c6a179\n",
      "2020-04-09 11:11:42,089 INFO Written out 2700000 reads\n",
      "2020-04-09 11:11:42,146 INFO Written out 1100000 reads\n",
      "2020-04-09 11:11:44,516 INFO Parsed 3000000 input reads\n",
      "2020-04-09 11:11:44,674 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:11:44,838 INFO Written out 2800000 reads\n",
      "2020-04-09 11:11:46,051 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:11:46,407 INFO Written out 1700000 reads\n",
      "2020-04-09 11:11:47,381 INFO Written out 1200000 reads\n",
      "2020-04-09 11:11:47,525 INFO Written out 1700000 reads\n",
      "2020-04-09 11:11:49,104 INFO Written out 2900000 reads\n",
      "2020-04-09 11:11:50,775 INFO Reads: Input Reads: 2321945\n",
      "2020-04-09 11:11:50,775 INFO Number of reads out: 1952649\n",
      "2020-04-09 11:11:50,775 INFO Total number of positions deduplicated: 1126886\n",
      "2020-04-09 11:11:50,775 INFO Mean number of unique UMIs per position: 2.02\n",
      "2020-04-09 11:11:50,775 INFO Max. number of unique UMIs per position: 7173\n",
      "# job finished in 168 seconds at Thu Apr  9 11:11:50 2020 -- 166.93  2.59  0.00  0.00 -- 27130992-ad40-4912-a014-5eddabccbfd2\n",
      "2020-04-09 11:11:56,352 INFO Reads: Input Reads: 1591375\n",
      "2020-04-09 11:11:56,352 INFO Number of reads out: 1097002\n",
      "2020-04-09 11:11:56,352 INFO Total number of positions deduplicated: 539450\n",
      "2020-04-09 11:11:56,352 INFO Mean number of unique UMIs per position: 2.74\n",
      "2020-04-09 11:11:56,352 INFO Max. number of unique UMIs per position: 11335\n",
      "# job finished in 174 seconds at Thu Apr  9 11:11:56 2020 -- 172.71  2.25  0.00  0.00 -- c6cce8d1-9524-45ae-b695-da061f47582b\n",
      "2020-04-09 11:11:56,645 INFO Written out 3000000 reads\n",
      "2020-04-09 11:11:58,154 INFO Written out 900000 reads\n",
      "2020-04-09 11:11:58,977 INFO Written out 1800000 reads\n",
      "2020-04-09 11:12:00,299 INFO Written out 1800000 reads\n",
      "2020-04-09 11:12:01,646 INFO Written out 3100000 reads\n",
      "2020-04-09 11:12:02,081 INFO Written out 200000 reads\n",
      "2020-04-09 11:12:04,126 INFO Written out 1000000 reads\n",
      "2020-04-09 11:12:08,362 INFO Written out 1300000 reads\n",
      "2020-04-09 11:12:12,350 INFO Written out 1400000 reads\n",
      "2020-04-09 11:12:16,263 INFO Written out 1500000 reads\n",
      "2020-04-09 11:12:17,001 INFO Written out 1100000 reads\n",
      "2020-04-09 11:12:17,320 INFO Written out 300000 reads\n",
      "2020-04-09 11:12:18,583 INFO Written out 1800000 reads\n",
      "2020-04-09 11:12:18,767 INFO Written out 1900000 reads\n",
      "2020-04-09 11:12:20,594 INFO Written out 1600000 reads\n",
      "2020-04-09 11:12:23,557 INFO Written out 2000000 reads\n",
      "2020-04-09 11:12:25,389 INFO Written out 3200000 reads\n",
      "2020-04-09 11:12:25,674 INFO Written out 1700000 reads\n",
      "2020-04-09 11:12:25,885 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:12:28,485 INFO Written out 1900000 reads\n",
      "2020-04-09 11:12:29,531 INFO Written out 1800000 reads\n",
      "2020-04-09 11:12:30,810 INFO Written out 1200000 reads\n",
      "2020-04-09 11:12:33,852 INFO Written out 3300000 reads\n",
      "2020-04-09 11:12:36,820 INFO Written out 1900000 reads\n",
      "2020-04-09 11:12:37,956 INFO Written out 2100000 reads\n",
      "2020-04-09 11:12:42,397 INFO Written out 2000000 reads\n",
      "2020-04-09 11:12:43,070 INFO Written out 400000 reads\n",
      "2020-04-09 11:12:43,900 INFO Written out 2000000 reads\n",
      "2020-04-09 11:12:44,337 INFO Written out 1300000 reads\n",
      "2020-04-09 11:12:46,693 INFO Written out 1900000 reads\n",
      "2020-04-09 11:12:53,555 INFO Reads: Input Reads: 2744526\n",
      "2020-04-09 11:12:53,555 INFO Number of reads out: 2170921\n",
      "2020-04-09 11:12:53,555 INFO Total number of positions deduplicated: 1265250\n",
      "2020-04-09 11:12:53,555 INFO Mean number of unique UMIs per position: 2.09\n",
      "2020-04-09 11:12:53,555 INFO Max. number of unique UMIs per position: 10623\n",
      "# job finished in 231 seconds at Thu Apr  9 11:12:53 2020 -- 228.38  3.84  0.00  0.00 -- c6b57d14-edee-4193-b996-4350a56b4444\n",
      "2020-04-09 11:12:55,793 INFO Reads: Input Reads: 2578507\n",
      "2020-04-09 11:12:55,793 INFO Number of reads out: 2074343\n",
      "2020-04-09 11:12:55,793 INFO Total number of positions deduplicated: 1104053\n",
      "2020-04-09 11:12:55,793 INFO Mean number of unique UMIs per position: 2.27\n",
      "2020-04-09 11:12:55,793 INFO Max. number of unique UMIs per position: 16903\n",
      "# job finished in 233 seconds at Thu Apr  9 11:12:55 2020 -- 231.15  3.58  0.00  0.00 -- cfec06b7-f88b-4c68-ad76-13cb79b08183\n",
      "2020-04-09 11:12:58,750 INFO Written out 3400000 reads\n",
      "2020-04-09 11:12:58,988 INFO Written out 2000000 reads\n",
      "2020-04-09 11:13:00,320 INFO Written out 1400000 reads\n",
      "2020-04-09 11:13:03,513 INFO Written out 2100000 reads\n",
      "2020-04-09 11:13:08,212 INFO Written out 3500000 reads\n",
      "2020-04-09 11:13:08,347 INFO Written out 2200000 reads\n",
      "2020-04-09 11:13:10,622 INFO Parsed 4000000 input reads\n",
      "2020-04-09 11:13:18,784 INFO Written out 2300000 reads\n",
      "2020-04-09 11:13:19,390 INFO Written out 500000 reads\n",
      "2020-04-09 11:13:23,108 INFO Written out 1500000 reads\n",
      "2020-04-09 11:13:24,133 INFO Written out 2100000 reads\n",
      "2020-04-09 11:13:24,494 INFO Reads: Input Reads: 4068818\n",
      "2020-04-09 11:13:24,495 INFO Number of reads out: 3574887\n",
      "2020-04-09 11:13:24,495 INFO Total number of positions deduplicated: 2178732\n",
      "2020-04-09 11:13:24,495 INFO Mean number of unique UMIs per position: 1.84\n",
      "2020-04-09 11:13:24,495 INFO Max. number of unique UMIs per position: 11717\n",
      "# job finished in 262 seconds at Thu Apr  9 11:13:24 2020 -- 258.08  4.97  0.00  0.00 -- fb1fe9c2-6852-43bb-9ee0-b00759c696dd\n",
      "2020-04-09 11:13:28,015 INFO Reads: Input Reads: 2812411\n",
      "2020-04-09 11:13:28,016 INFO Number of reads out: 2313216\n",
      "2020-04-09 11:13:28,016 INFO Total number of positions deduplicated: 1416021\n",
      "2020-04-09 11:13:28,016 INFO Mean number of unique UMIs per position: 1.93\n",
      "2020-04-09 11:13:28,016 INFO Max. number of unique UMIs per position: 18329\n",
      "# job finished in 266 seconds at Thu Apr  9 11:13:28 2020 -- 262.53  4.16  0.00  0.00 -- 6f7b3750-a7c7-4ade-8dac-266b1ce46bc8\n",
      "2020-04-09 11:13:31,746 INFO Parsed 2000000 input reads\n",
      "2020-04-09 11:13:35,204 INFO Written out 2200000 reads\n",
      "2020-04-09 11:13:40,075 INFO Written out 200000 reads\n",
      "2020-04-09 11:13:41,072 INFO Written out 1600000 reads\n",
      "2020-04-09 11:13:44,351 INFO Reads: Input Reads: 2819734\n",
      "2020-04-09 11:13:44,352 INFO Number of reads out: 2238192\n",
      "2020-04-09 11:13:44,352 INFO Total number of positions deduplicated: 1135861\n",
      "2020-04-09 11:13:44,352 INFO Mean number of unique UMIs per position: 2.42\n",
      "2020-04-09 11:13:44,352 INFO Max. number of unique UMIs per position: 16497\n",
      "# job finished in 282 seconds at Thu Apr  9 11:13:44 2020 -- 278.35  4.90  0.00  0.00 -- 4bb1dee8-f8d7-4645-918b-400d1971d433\n",
      "2020-04-09 11:13:47,456 INFO Written out 1700000 reads\n",
      "2020-04-09 11:13:49,517 INFO Written out 300000 reads\n",
      "2020-04-09 11:14:01,839 INFO Parsed 1000000 input reads\n",
      "2020-04-09 11:14:06,020 INFO Written out 400000 reads\n",
      "2020-04-09 11:14:14,937 INFO Written out 600000 reads\n",
      "2020-04-09 11:14:16,146 INFO Written out 1800000 reads\n",
      "2020-04-09 11:14:17,656 INFO Reads: Input Reads: 791307\n",
      "2020-04-09 11:14:17,656 INFO Number of reads out: 458769\n",
      "2020-04-09 11:14:17,656 INFO Total number of positions deduplicated: 117053\n",
      "2020-04-09 11:14:17,656 INFO Mean number of unique UMIs per position: 6.13\n",
      "2020-04-09 11:14:17,656 INFO Max. number of unique UMIs per position: 30185\n",
      "# job finished in 315 seconds at Thu Apr  9 11:14:17 2020 -- 312.65  3.71  0.00  0.00 -- f318f2d2-e3a6-4a72-a9e9-2df99c3f10a8\n",
      "2020-04-09 11:14:32,318 INFO Reads: Input Reads: 2501926\n",
      "2020-04-09 11:14:32,318 INFO Number of reads out: 1836040\n",
      "2020-04-09 11:14:32,318 INFO Total number of positions deduplicated: 976658\n",
      "2020-04-09 11:14:32,318 INFO Mean number of unique UMIs per position: 2.45\n",
      "2020-04-09 11:14:32,318 INFO Max. number of unique UMIs per position: 10696\n",
      "# job finished in 330 seconds at Thu Apr  9 11:14:32 2020 -- 324.33  6.65  0.00  0.00 -- 8fac16bd-f125-448c-b1ea-21f3f3cd3c8d\n",
      "2020-04-09 11:14:47,895 INFO Reads: Input Reads: 1124450\n",
      "2020-04-09 11:14:47,895 INFO Number of reads out: 665732\n",
      "2020-04-09 11:14:47,895 INFO Total number of positions deduplicated: 123393\n",
      "2020-04-09 11:14:47,895 INFO Mean number of unique UMIs per position: 8.51\n",
      "2020-04-09 11:14:47,895 INFO Max. number of unique UMIs per position: 21060\n",
      "# job finished in 346 seconds at Thu Apr  9 11:14:47 2020 -- 341.66  4.89  0.00  0.00 -- 0612eb08-9c68-48c9-aae4-c2b6104ab138\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for f in *.srt.bam; do\n",
    "out=${f/.srt.bam/.srt.dd.bam};\n",
    "echo \"umi_tools dedup -I $f -S $out\";\n",
    "umi_tools dedup -I $f -S $out &\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools sort -o S01.sort.dd.bam S01.srt.dd.bam\n",
      "samtools sort -o S02.sort.dd.bam S02.srt.dd.bam\n",
      "samtools sort -o S03.sort.dd.bam S03.srt.dd.bam\n",
      "samtools sort -o S04.sort.dd.bam S04.srt.dd.bam\n",
      "samtools sort -o S05.sort.dd.bam S05.srt.dd.bam\n",
      "samtools sort -o S06.sort.dd.bam S06.srt.dd.bam\n",
      "samtools sort -o S07.sort.dd.bam S07.srt.dd.bam\n",
      "samtools sort -o S08.sort.dd.bam S08.srt.dd.bam\n",
      "samtools sort -o S09.sort.dd.bam S09.srt.dd.bam\n",
      "samtools sort -o S10.sort.dd.bam S10.srt.dd.bam\n",
      "samtools sort -o S11.sort.dd.bam S11.srt.dd.bam\n",
      "samtools sort -o S12.sort.dd.bam S12.srt.dd.bam\n",
      "samtools sort -o S13.sort.dd.bam S13.srt.dd.bam\n",
      "samtools sort -o S14.sort.dd.bam S14.srt.dd.bam\n",
      "samtools sort -o S15.sort.dd.bam S15.srt.dd.bam\n",
      "samtools sort -o S16.sort.dd.bam S16.srt.dd.bam\n",
      "samtools sort -o S17.sort.dd.bam S17.srt.dd.bam\n",
      "samtools sort -o S18.sort.dd.bam S18.srt.dd.bam\n",
      "samtools sort -o S19.sort.dd.bam S19.srt.dd.bam\n",
      "samtools sort -o S20.sort.dd.bam S20.srt.dd.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 4 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for f in *.srt.dd.bam; do\n",
    "out=${f/.srt.dd.bam/.sort.dd.bam};\n",
    "echo \"samtools sort -o $out $f\";\n",
    "samtools sort -@ 4 -m 2G -o $out $f\n",
    "samtools index $out;\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processInput(path):\n",
    "    samfile = pysam.AlignmentFile(path, \"rb\")\n",
    "    data = samfile.fetch()\n",
    "    sample_name = re.split(\"\\.\",re.split(\"/\", path)[-1])[0]\n",
    "    documents = {}\n",
    "    for read in data:\n",
    "        mapflag = False\n",
    "        mapq = read.mapping_quality\n",
    "        if mapq<15:\n",
    "            mapflag = True\n",
    "        seq = read.get_forward_sequence()\n",
    "        ref_id = read.reference_name\n",
    "        ref_start = read.reference_start\n",
    "        ref_end = read.reference_end\n",
    "        strand = \"+\"\n",
    "        if read.is_reverse:\n",
    "            strand = \"-\"\n",
    "        if seq in documents and documents[seq][\"strand\"] == strand:\n",
    "            seq_dict = documents[seq]\n",
    "            seq_dict[\"count\"] += 1\n",
    "            if mapflag and not seq_dict[\"mapflag\"]:\n",
    "                seq_dict[\"mapflag\"] = mapflag\n",
    "        else:\n",
    "            documents[seq] = {\"seq\": seq, \"count\":1, \n",
    "                             \"ref_id\": ref_id, \"start\": ref_start,\n",
    "                              \"end\": ref_end, \"strand\": strand,\n",
    "                              \"len\" : len(seq),\n",
    "                              \"mapflag\": mapflag}\n",
    "    df = pd.DataFrame.from_dict(documents,orient=\"index\")\n",
    "    df = df.sort_values(by=\"count\", ascending=False)\n",
    "    df.index = pd.RangeIndex(len(df.index))\n",
    "    df.to_csv(f\"{sample_name}_data.csv\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "path=\"./\"\n",
    "bamfiles = []\n",
    "for subf in os.scandir(path):\n",
    "    if subf.name.endswith(\".sort.dd.bam\"):\n",
    "        bamfiles.append(subf.path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_cores = cpu_count()\n",
    "results = Parallel(n_jobs=num_cores)(delayed(processInput)(i) for i in bamfiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
