{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Breast Cancer Cell Lines: Preprocess Data\n",
    "In this notebook we preprocess our data for downstream locus-based analysis. Data includes:\n",
    "* smRNA seq data of the breast cancer (and normal) cell lines used in our Nature Med paper (Fish, 2018)\n",
    "* smRNA seq data from TCGA: specifically all the BRCA samples and all the normal samples from all tissue types in TCGA. \n",
    "* smRNA seq data from exRNA Atlas along with other publically available datasets from non-cancerous exosomal smRNA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pysam\n",
    "import re\n",
    "import json as js\n",
    "from scipy.signal import find_peaks, peak_widths\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Cell Lines Data\n",
    "Dataset involves 21 samples:\n",
    "* 3 samples from HMEC cell lines\n",
    "* 18 samples from "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_lines = [f for f in os.scandir(\"data/IC/\") if f.name.endswith(\".srt.dd.bam\")]\n",
    "len(cell_lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dust"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we are converting the bam files to bed format, we first remove the low complexity sequences in the bam files first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Simple dust score adapted from Dust Algorithm Score: A fast and symmetric DUST implementation to mask\n",
    "low-complexity DNA sequences (2006, Morgulis et al.)\"\"\"\n",
    "def simpleDustScore(seq):\n",
    "    assert len(seq) > 2\n",
    "    if len(seq) == 3:\n",
    "        return 0\n",
    "    else:\n",
    "        triplets = {}\n",
    "        num_trip = len(seq) - 2\n",
    "        for i in range(num_trip):\n",
    "            subseq = seq[i:i+3]\n",
    "            if subseq in triplets:\n",
    "                triplets[subseq] += 1\n",
    "            else:\n",
    "                triplets[subseq] = 1\n",
    "        sum_triplet = 0\n",
    "        for triplet, count in triplets.items():\n",
    "            sum_triplet += count * (count - 1) / 2\n",
    "        return sum_triplet/(num_trip - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in cell_lines: \n",
    "    name = f.name.split(\".\")[0]\n",
    "    out = f\"data/IC/{name}.srt.dd.dust.bam\"\n",
    "    infile = pysam.AlignmentFile(f, \"rb\")\n",
    "    outfile = pysam.AlignmentFile(out, \"wb\", template=infile)\n",
    "    for read in infile.fetch():\n",
    "        if simpleDustScore(read.get_forward_sequence()) < 3 and len(read.get_forward_sequence()) >= 15:\n",
    "            outfile.write(read)\n",
    "    outfile.close()\n",
    "    infile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BedtoBam\n",
    "Next we convert all the cell lines bam alignment files to bedfiles. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for f in data/IC/*.srt.dd.dust.bam;\n",
    "do \n",
    "base=$(basename $f)\n",
    "out=${base/.bam/.bed}\n",
    "echo \"$bedtools bamtobed -i $f > data/IC/$out\"\n",
    "bedtools bamtobed -i $f > data/IC/$out\n",
    "done &> log/cell_lines_bamtobed.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter Cell Lines Bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in os.scandir(\"data/IC\"):\n",
    "    if f.name.endswith(\".srt.dd.dust.bed\"):\n",
    "        name = f.name.split(\".\")[0]\n",
    "        outfile = f\"data/IC/{name}.filter.bed\"\n",
    "        with open(outfile, \"wt\") as out, open(f, \"rt\") as file:\n",
    "            for line in file:\n",
    "                if re.match(\"chr[\\d+,X,Y]\", line) and \"None\" not in line:\n",
    "                    out.write(line) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge Cell Lines Loci\n",
    "Here we merge all the loci seen in the cell lines data to create a locus feature map. No threshold set for degree of overlap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat data/IC/*.filter.bed | sort -k1,1 -k2,2n | mergeBed -s -c 6 -o distinct -i stdin | \\\n",
    "awk '{print $1 \"\\t\" $2 \"\\t\" $3 \"\\t\" $1\":\"$2\"-\"$3\":\"$4  \"\\t.\" \"\\t\" $4  > \"data/cell_lines_smRNAs_merged_loci.bed\"}' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split Loci By Peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since some loci are larger than 200 bp, we will need to split these loci up by coverage. <br>\n",
    "First we merge all the cell lines sample bamfiles to run `bedtools genomecov`. Based on coverage peaks we will separate the larger loci into smaller ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "samtools merge data/IC/all_samples.bam data/IC/*srt.dd.dust.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools genomecov -ibam data/IC/all_samples.bam -bg -strand - > data/IC/all_samples_neg_bg.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools genomecov -ibam data/IC/all_samples.bam -bg -strand + > data/IC/all_samples_pos_bg.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629083</td>\n",
       "      <td>630075</td>\n",
       "      <td>chr1:629083-630075:+</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629218</td>\n",
       "      <td>629905</td>\n",
       "      <td>chr1:629218-629905:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630084</td>\n",
       "      <td>630520</td>\n",
       "      <td>chr1:630084-630520:+</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630317</td>\n",
       "      <td>630520</td>\n",
       "      <td>chr1:630317-630520:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630645</td>\n",
       "      <td>631429</td>\n",
       "      <td>chr1:630645-631429:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0       1       2                     3  4  5  length\n",
       "224  chr1  629083  630075  chr1:629083-630075:+  .  +     992\n",
       "225  chr1  629218  629905  chr1:629218-629905:-  .  -     687\n",
       "228  chr1  630084  630520  chr1:630084-630520:+  .  +     436\n",
       "231  chr1  630317  630520  chr1:630317-630520:-  .  -     203\n",
       "237  chr1  630645  631429  chr1:630645-631429:-  .  -     784"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_loci = pd.read_csv(\"data/cell_lines_smRNAs_merged_loci.bed\", header=None, sep=\"\\t\")\n",
    "all_loci[\"length\"] = all_loci[2] - all_loci[1] #Calculate length of each locus\n",
    "loci_to_split = all_loci[all_loci[\"length\"] > 200]\n",
    "loci_to_split.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11046</td>\n",
       "      <td>11061</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11237</td>\n",
       "      <td>11258</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11768</td>\n",
       "      <td>11785</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12036</td>\n",
       "      <td>12072</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17017</td>\n",
       "      <td>17035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2  3\n",
       "0  chr1  11046  11061  1\n",
       "1  chr1  11237  11258  1\n",
       "2  chr1  11768  11785  1\n",
       "3  chr1  12036  12072  1\n",
       "4  chr1  17017  17035  1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_density = pd.read_csv(\"data/IC/all_samples_pos_bg.txt\", header=None, sep=\"\\t\")\n",
    "neg_density = pd.read_csv(\"data/IC/all_samples_neg_bg.txt\", header=None, sep=\"\\t\")\n",
    "pos_density.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4400, 1011)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_loci_list = []\n",
    "for i in range(loci_to_split.shape[0]):\n",
    "    rna = loci_to_split.iloc[i]\n",
    "    density = neg_density if rna[5] == \"-\" else pos_density\n",
    "    density = density[density[0] == rna[0]] #Match reference chr\n",
    "    density = density[(density[1] >= rna[1]) & (density[2] <= rna[2])] #Match rna locus. Since the loci area already merged, the coverage start and end positions should match seamlessly.\n",
    "    coord, coverage = [], []\n",
    "    assert density.iloc[0][1] == rna[1]\n",
    "    assert density.iloc[-1][2] == rna[2] \n",
    "    posit = density.iloc[0][1] #Start of locus\n",
    "    for i in range(density.shape[0]):\n",
    "        row = density.iloc[i]\n",
    "        start, end, count = row[1], row[2], row[3]\n",
    "        \n",
    "        #Mergebed and genomecov results have about 6 loci that are off by 1~3 bps. Here we fix continuity  \n",
    "        while posit < start:\n",
    "            coord.append(posit)\n",
    "            coverage.append(0) #Bridge gap in genomecov with 0s\n",
    "            posit += 1\n",
    "            \n",
    "        #Generating coverage by each nucletotide position.\n",
    "        while start < end:\n",
    "            coord.append(start)\n",
    "            coverage.append(count)\n",
    "            start += 1\n",
    "            posit += 1\n",
    "            \n",
    "    peaks, _ = find_peaks(np.array(coverage), height=5, rel_height=0.5, width=[15, 200], distance=20)\n",
    "    results = peak_widths(np.array(coverage), peaks, rel_height=0.5)\n",
    "    #peak_widths returns 4 arrays in order: peak widths, peak heights, peak start, peak end.\n",
    "    for new_start_index,new_end_index in zip(results[2], results[3]):\n",
    "        new_start_index, new_end_index = int(new_start_index), int(new_end_index) #These are the new indices\n",
    "        new_start = coord[new_start_index]\n",
    "        new_end = coord[new_end_index]\n",
    "        \n",
    "        if new_end - new_start + 7 < 200: #Add an additional 6 nt buffer if loci length is not too large.\n",
    "            new_start -= 3\n",
    "            new_end += 4 #1 nt more for end because bed format is end exclusive.\n",
    "        new_locus = f\"{rna[0]}:{new_start}-{new_end}:{rna[5]}\"\n",
    "        new_loci_list.append(new_locus)\n",
    "        \n",
    "len(new_loci_list), loci_to_split.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the 1011 loci greater than 200 bp, we split them into 4400 loci."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1734531"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_list = all_loci[all_loci[\"length\"] <= 200][3].tolist()\n",
    "loci_list.extend(new_loci_list)\n",
    "len(loci_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/cell_lines_smRNAs_split_loci.bed\", \"wt\") as out:   \n",
    "    for locus in loci_list:\n",
    "        splits = locus.split(\":\")\n",
    "        start = splits[1].split(\"-\")[0]\n",
    "        end = splits[1].split(\"-\")[1]\n",
    "        bed = f\"{splits[0]}\\t{start}\\t{end}\\t{locus}\\t.\\t{splits[-1]}\"\n",
    "        out.write(bed + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat data/cell_lines_smRNAs_split_loci.bed | sort -k1,1 -k2,2n | mergeBed -s -c 6 -o distinct -i stdin | \\\n",
    "awk '{print $1 \"\\t\" $2 \"\\t\" $3 \"\\t\" $1\":\"$2\"-\"$3\":\"$4  \"\\t.\" \"\\t\" $4  > \"data/cell_lines_smRNAs_loci_map.bed\"}' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1734299"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_map = pd.read_csv(\"data/cell_lines_smRNAs_loci_map.bed\", header=None, sep=\"\\t\")[3]\n",
    "len(loci_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Final number of loci in our feature map is 1734299."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## exRNA Filter Feature Map\n",
    "Here we filter out all the loci that are also seen in non-cancerous samples from the exRNA Atlas and other publically available non-cancerous exosomal smRNAseq datasets. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command: <br>\n",
    "`ls /rumi/shams/jwang/exRNA_atlas/data/bedfiles/*.srt.filter.bed | parallel -j 30 bash scripts/intersect_exRNA_atlas_loci.sh {} &> log/intersect_exRNA_atlas_loci.out` <br>\n",
    "\n",
    "\n",
    "The main command in the script is: <br>\n",
    "`intersectBed -s -u -a data/cell_lines_smRNAs_loci_map.bed -b $1 > data/exRNA/$out` <br>\n",
    "The `-u` option reports a locus in `-a` **once** when there is at least one overlap with `-b`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot exRNA clean up threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1383"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_intersect_files = [f for f in os.scandir(\"data/exRNA\") if os.stat(f).st_size != 0 and f.name.endswith(\".bed\")] #Exclude empty no hit samples.\n",
    "len(exRNA_intersect_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1143299"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_loci_hits = {}\n",
    "for file in exRNA_intersect_files:\n",
    "    df = pd.read_csv(file, sep=\"\\t\", header=None)\n",
    "    for locus in df[3]:\n",
    "        if locus in exRNA_loci_hits:\n",
    "            exRNA_loci_hits[locus] += 1\n",
    "        else:\n",
    "            exRNA_loci_hits[locus] = 1\n",
    "len(exRNA_loci_hits) #Represents number of loci in our loci feature map that had at least one overlap with a sample in the exRNA atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>loci</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:629497-629544:+</td>\n",
       "      <td>1098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:629535-629573:-</td>\n",
       "      <td>1092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:629567-629614:+</td>\n",
       "      <td>1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:630746-630788:+</td>\n",
       "      <td>1027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:630755-630794:-</td>\n",
       "      <td>799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   loci  counts\n",
       "0  chr1:629497-629544:+    1098\n",
       "1  chr1:629535-629573:-    1092\n",
       "2  chr1:629567-629614:+    1187\n",
       "3  chr1:630746-630788:+    1027\n",
       "4  chr1:630755-630794:-     799"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_loci_hits_df = pd.DataFrame.from_dict(exRNA_loci_hits, orient=\"index\", columns=[\"counts\"])\n",
    "exRNA_loci_hits_df = exRNA_loci_hits_df.reset_index(drop=False).rename(columns={\"index\":\"loci\"})\n",
    "exRNA_loci_hits_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_loci_num = 1734299\n",
    "exRNA_filter_results = {\"threshold\":[], \"count\":[]}\n",
    "for th in range(1, 50):\n",
    "    exRNA_filter_results[\"threshold\"].append(th)\n",
    "    exRNA_filter_results[\"count\"].append(total_loci_num - np.sum(exRNA_loci_hits_df[\"counts\"] >= th)) #Number of total loci minus number of loci eliminated (loci that intersected with or more than th samples) given a threshold TH\n",
    "exRNA_filter_results = pd.DataFrame(exRNA_filter_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAERCAYAAAB2CKBkAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAxrElEQVR4nO3deXxU9dX48c8h+0aAJKwBw66AiEoRxQpqtSq41PK4PE/dFa31sf7aau2itk+1i1upUhequLYite7VugNqRQXFCkEQESUQkhBIJiGZrOf3x73BgMnkZpm5mZnzfr3mdecuc++5+uKe3O8qqooxxpj41cfvAIwxxvjLEoExxsQ5SwTGGBPnLBEYY0ycs0RgjDFxzhKBMcbEuahMBCKySERKRWSNx+PPEJFCEVkrIn8Ld3zGGBNNJBr7EYjIUUA18LCqTurg2LHAEuAYVd0lIgNVtTQScRpjTDSIyjcCVV0O7Gy9TURGi8i/RGSViLwpIvu7uy4B/qyqu9zfWhIwxphWojIRtGMh8L+qeijwE+Aud/s4YJyIvC0iK0TkBN8iNMaYXijR7wB6gohkAkcAfxeRls0p7jIRGAvMAvKBN0VkkqpWRDhMY4zplWIiEeC82VSo6pQ29hUBK1S1AfhcRNbjJIb3IxifMcb0WjFRNKSqAZyH/H8BiOMgd/fTwNHu9lycoqJNfsRpjDG9UVQmAhF5DHgHGC8iRSJyEfA/wEUi8hGwFjjVPfwloFxECoE3gKtVtdyPuI0xpjeKyuajxhhjek5UvhEYY4zpOVFXWZybm6sFBQV+h9E7rV/vLMeP9zcOY0yvs2rVqh2qmtfWvqhLBAUFBaxcudLvMHqnWbOc5dKlfkZhjOmFROSL9vZZ0ZAxxsS5qHsjMCH88pd+R2CMiUKWCGLJt77ldwTGmChkRUOxZPVq52OMMZ1gbwSx5KqrnKVVFhtjOsHeCIwxJs5ZIjDGmDhnRUPGGOOD5mYl2NhETX0TtfVN1DY432vqGwk2fLV9z/eGJg7drz/fHNtmn7BusURgjDHtUFVqG5rYXec8oFuWLQ/s3XVN1DQ0UVPXuOdh3bK/tn7fh3wjwYbmvb531vdnjbZEYDrw29/6HYExvmtsaqa6rpGqYCPVde4n2EhVXSO7W32vDjZSU+/sr6lvotrdX1Pf9NWyvpHOjMuZmtSH9ORE0pISyEhJIC05kfSkBPKyUkhLTictKYG0pATSkxNIdZdffU/ca3tasnNsWssxiQn06SMdB9EFlghiyRFH+B2BMV2mqgQbmgkEG6gKNlLlLp2Hesu2xj37Wh72VcEGqloe/MFGahuaPF0vIzmBjJREMlMSSU9JICM5kcF9U0lPSSQzxXkwZ6QkkpGcQHrLMrllewJpSe7S3ZaeFL4HdbhZIogl//63s7SEYHxS39hMRW09gdoGKmqcT2VtA4FgA4HaRne593rrh35jc8d/fmckJ5CVmkRWaiJZqYlkpyeTPyCdvqnOQz0zxdmXmZpIVorzMM9q2ecuM5ITo/ahHQ6WCGLJz3/uLK0fgemmusYmKmsa2FXTQEVNPbtqGqisrXfXne8tD/qK2gYqa+qpqG2gpj70X+MZyQn0TUsiO815WA/qm8qYgYnuQz1pz7Jv6lfbMlP2/p5gD/AeZ4nAmBinqlTXNVJeXU/57np27q6nvLqO8t317HLXd9Y431u27Q7xQE9KEPqlJ9MvLYl+6UkM65fGxKF9yU5Lon+685DPdvdnt/pkpSaSmGAt1nsjSwTGRCFVJVDbSFl1HWVVdXuWO6rr2OEuy3fXO99311Pf2HYLlbSkBAZkJO/5jMrLpH96MgMykpyHfXoS/dOTnYd8hvNwT09OQMT+Ko8lYUsEIrIImAOUquqkdo6ZBcwHkoAdqjozXPEYEw2ampUd1XVsrwxSWlVHaVWQsqo653vAeeDvqHIe+vVNX3+4JyUIORkp5GYlk5ORwtiBWeRmJZObkeI87DOTyXEf+jkZKaQlJ/hwl6a3CecbwYPAAuDhtnaKSD/gLuAEVf1SRAaGMRZjfFff2ExJIMj2QJDiyiDFFbUUVwYpcddLAs7Dv6mNCtMBGckMzEohLyuF0bkZ5PVNIS/TWc/LSmFgVgq5mSlkpyXZX+um08KWCFR1uYgUhDjkv4EnVfVL9/jScMUSN+bP9zuCuFZd10jRrhq27qqlaFctWytq3e81bKsMsqO67mtt0jNTEhmcncrgvqmMGZPL4L6pDHLXB/V1HvK5mSkkWdm6CSM/6wjGAUkishTIAv6kqu29PcwD5gGMGDEiYgFGnSlT/I4gpjU0NbOtopYvymvYsquGL3fWULSzli27atiys4ZdNQ17HZ+c2If8fmkM65/G/oP7MqRfKkOz0xicncqQ7FQGZ6eSlZrk090Y8xU/E0EicChwLJAGvCMiK1R1w74HqupCYCHA1KlTO9HPL868+qqztAlquqyhqZktO2vYVLabzeW7+aK8hs3lu52H/q7avYptkhP6MKx/Gvn905h04BCG909n+IA0hrkP/9yMFGurbqKCp0QgIoOBaYAC76vq9h64dhFOBfFuYLeILAcOAr6WCIxHN97oLC0RdKiytoGNpdV8VlrNZ2XVfFa2m007qvmyvGavTk1ZqYmMzM3gwGHZnDx5KPvlpDNiQDojctIZlJVqD3oTEzpMBCJyMXA98DogwJ0i8n+quqib134GWCAiiUAycBjwx26e05i9VNTUs357FRtKqthYWs2npdVsLK2mtKpuzzHJCX3YLyedsQMz+fbEwYzKzWBUXiajcjPol26Vryb2eXkjuBo4WFXLAUQkB/g3EDIRiMhjwCwgV0SKgBtwmomiqveo6joR+RfwH6AZuE9V13T1Rkx8q2ts4tOSatYVB9hQUsUn7sO/JPDVAz8zJZHRAzM5alweYwZmMnZgJmMGZpLfP916q5q45iURFAFVrdargC0d/UhVz/ZwzC3ALR5iMGaPytoGCrcFKCwOsHZbJYXbAmwsrd5TpJOc2IexAzOZMSaX8YOyGD84i3GDshiSnWp/3RvTBi+JYCvwrog8g1NHcCrwnoj8CEBVbw9jfCbO1TU2sa64itVf7uKjokpWb6ng8x279+wfmJXChKF9OWb/gUwcms3+Q7IoyMmwv/CN6QQvieAz99PiGXeZ1fPhmG65916/I+i2bRW1rPpiF6u+2MWHWypYty2wpwftwKwUpgzvx9xD85k0LJsJQ/qSl5Xic8TGRL8OE4Gq/joSgZgeMH683xF0SlOzUrgtwPubd7Lqy1188MUuiiuDgDPBx0H5/bjgyAKm5Pdjyoh+DO5rRTvGhIOXVkN5wDXARCC1ZbuqHhPGuExXPPecszz5ZH/jaEdzs1JYHGDFpnJWbCrn3c93UhVsBGBYvzSmFgzg0BH9OHS/Aew/JMt60xoTIV6Khv4KPI4zgNxlwHlAWTiDMl10223Oshclgq0VtSxdX8ryDWWs2LSTylqn921BTjpzJg9h+qgcpo0cwJDsNJ8jNSZ+eUkEOap6v4j8UFWXActEZFm4AzPRqa6xifc/38XS9aUs3VDGxtJqwPmL//gJgzh8dA6Hj86xB78xvYiXRNAygEqxiMwGtgH54QvJRJuKmnpeW1fKy4XbefPTHdTUN5Gc0IdpIwdw1jeGM2t8HqPzMq1835heyksiuFFEsoEfA3cCfYH/F9aoTK+3raKWl9du5+XCEt79fCdNzcrgvqmcfsgwjh4/kMNH55CebPMeGRMNvLQaet79WgkcHd5wTG9WWhXkuY+KeXb1Vj4qqgRgzMBMLps5iuMnDGZyfrb91W9MFGo3EYjInTgdyNqkqleGJSLTdY880uOnrAo28NLaEp5ZvZW3N+6gWWHi0L789IT9OX7iIEbnZfb4NY0xkRXqjWBlxKIwPWP48B45TXOz8tbGHTy+cguvFpZQ19jM8AFpXD5rDKcdPJQxA60voTGxpN1EoKoPiUgC8HtVvTqCMZmuevxxZ3nmmV36eUkgyN9XbmHx+1so2lVLv/Qkzpg6nNMOHsohI/pbsY8xMSpkHYGqNonIoZEKxnTT3Xc7y04kgqZmZdmGUh57bwuvf1JKU7NyxOgcrjlhf749cRApiTa5uTGxzkuzjg9F5Fng78Ce0b5U9cmwRWXCLtjQxBOrivjLm5v4oryG3MxkLvnmKM76xnAKcjP8Ds8YE0FeEsEAoBxoPaSEApYIolBlbQOPrviCB97ezI7qOg7Kz+bq/z6Y4ycMJjnRhnQwJh55aT56QSQCMeFVEghy/1uf87d3v6S6rpGjxuVx2cxRHD4qx8r+jYlzXgadGwfcDQxS1UkiMhk4RVVvDHt0ptsqaxu4a+lGHnx7Mw1NzcyePJRLjxrFpGHZfodmjOklvBQN/QVnusp7AVT1PyLyN8ASQW/zxBN7vgYbmnjknS9Y8MZGAsEGvjNlGFd9axwjctJ9DNAY0xt5SQTpqvrePsUHjWGKx3RHbi7NzcrTHxRx28sb2FpRy1Hj8rj2hP2ZMLSv39EZY3opL4lgh4iMxu1lLCJzgeKwRmW6ZPOtC/jHB9u4c8QMJg3ry81zJzNjTK7fYRljejkvieAHwEJgfxHZCnwO/E9YozKdUhVs4OZ/rWf2HfdydGICY675ASdPHkofm7fXGOOBl0Twhap+S0QygD6qWhXuoIx3L63dzg3PrKW0KshF2WkM75/GIVOG+R2WMSaKeGk4/rmILASmA9Vhjsd4VBIIctkjq7j0kVX0S0/iqctnUJCTToK9BRhjOsnLG8F44GScIqL7ReR5YLGqvhXWyEy7/rGqiF89t5b6xmauOWE8l3xzlM3va4zpMi8dymqBJcASEekP/AlYBtggNBFWW9/E9c+s4e+ripg2cgA3f3eyDQdhjOk2T1NIichM4EzgROB94IxwBmW+bmNpNT/46wdsKK3iymPH8sNjx369GOiFF/wJzhgT1bz0LP4cWI3zVnC1qu4O/QvT055ZvZWfPfkxaUkJPHTBNI4al9f2genWWcwY03le3ggOUtVA2CMxXxNsaOLXzxXy2Htf8o2C/tx59iEMzk5t/wd33eUsL788MgEaY2JCyBpGETkReF5EdohImYgsE5GTvJxYRBaJSKmIrOnguG+ISJPbUc24dlTX8V/3vMNj733JZTNH89gl00MnAYAlS5yPMcZ0Qqg5iy8BLgWu4atpK6cCvxeRfFVd2MG5HwQWAA+HuEYC8AfgpU7EHPO2VtRyzn3vsq2ylr+cO5XjJgzyOyRjTAwLVTT0/4AjVXVnq22vu28Jb+H0Nm6Xqi4XkYIOrv+/wD+Ab3iINS58VlbNOfe9S1VdI49edBhTCwb4HZIxJsaFKhqSfZIAAKpa3hMXFpFhwHeAezwcO09EVorIyrKysp64fK+0ZmslZ9zzDvVNzSyeN92SgDEmIkIlgoCIHLTvRndbTwwzMR/4qao2dXSgqi5U1amqOjUvr50WM1Hu/c07OXvhClIS+7Dk0sOZONTmCzDGREaooqEfA8+KyAPAKpzRR78BnAd8rweuPRVY7A5vnQucJCKNqvp0D5w7qixdX8plj65iaHYaj1x8GMP6pXXxREt7NC5jTHxoNxGo6lsiMg1naInzAQHWAtNVdXt3L6yqI1u+i8iDwPPxmASWbyjjkodXMnZgFg9fNI3czBS/QzLGxJmQ/QhUtQS4visnFpHHgFlArogUATcASe55O6wXiAefllTxg79+wOi8TB6bN53stKTunfDWW53lT37S/eCMMXHD0xATXaGqZ3fi2PPDFUdvVV5dx4UPvU9KUgL3n/+N7icBgOefd5aWCIwxnWBDVvqgrrGJyx5dRWmgjr+ce2jX6wSMMaYHdCoRiEgfEbHJb7tBVfnZkx/z/uZd3HbGQRw8or/fIRlj4lyHiUBE/iYifd0ZygqB9SJydfhDi013L/uMJz/Yyv/71jjmTB7qdzjGGOPpjWCCO+jcacALwAjgnHAGFav+taaYm/+1nlMOGsqVx47p+QukpTkfY4zpBC+VxUkikoSTCBaoaoOIaHjDij1rtlZy1eOrOWREP26eOxm3/0TPevHFnj+nMSbmeXkjuBfYDGQAy0VkP8CGpe6EYEMTVz72IQPSk7n3nKmkJtnkbsaY3sPLVJV3AHe02vSFiBwdvpBiz5/f2MimHbt55KJp5GWFscPYb37jLK+7LnzXMMbEHC+VxYNE5H4RedFdn4AzzITx4JPtAe5e+hmnHzKMb44N8zhJr73mfIwxphO8FA09iDNfQEsTlw3AVWGKJ6Y0NSvX/uNj+qYl8cvZE/wOxxhj2uQlEeSq6hKgGUBVG4EORww18Mg7m1m9pYLr50xgQEay3+EYY0ybvCSC3SKSgzP6KCIyHagMa1QxYFtFLbe8tJ6jxuVx6hTrL2CM6b28NB/9EfAsMFpE3gbyAJtfOARV5bqn19CscNNpk8LTVLQtOTmRuY4xJqZ4aTX0gYjMBMbjDEW9XlUbwh5ZFHv+P8W89kkpv5x9AMMHpEfuwv/4R+SuZYyJGaEmrz+9nV3jRARVfTJMMUW1ipp6fv3cWibnZ3PBjJEd/8AYY3wW6o3g5BD7FLBE0IbfvrCOXTUNPHzhYST0iVCRUIuf/cxZ/u53kb2uMSaqhZqh7IJIBhILNpZWs2RlEfOOGsWEoT4M0vrOO5G/pjEm6nnpUJYjIneIyAciskpE/uS2IjL7+MvyTaQk9uHSo0b5HYoxxnjmpfnoYqAM+C5Oa6Ey4PFwBhWNSgNBnvpwK2dMHU6OzTtsjIkiXpqPDlDV37Rav1FETgtTPFHrgX9vprG5mYu/aRXExpjo4uWN4A0ROcudnayPiJwB/DPcgUWTqmADj674ghMnDWG/nAz/AsnPdz7GGNMJXt4ILsXpVPaou94Hp7fxjwBV1bifunLxe1uoCjYyz++6gUcf7fgYY4zZh5cOZVmRCCRa1Tc2c/9bn3P4qBwOGt7P73CMMabTvLwRICKTgYLWx1uHMsezH21jeyDI7797oN+hwFVXOcv58/2MwhgTZTpMBCKyCJgMrMUdgRTrUAZAc7OycPln7D84i5njwjzXgBerV/sdgTEmCnl5I5iuqjaYfhuWbihlQ0k1fzzzoMgNLGeMMT3MS6uhd9xZycw+7lm2iaHZqcyZbMNMG2Oil5c3godwksF2oA5nBFJV1clhjayX++DLXbz3+U5+OfsAkhK85FNjjOmdvCSCRcA5wMd8VUcQ9xYu20Tf1ETOmjbC71C+Mm6c3xEYY6KQl0Twpao+29kTu5XMc4BSVZ3Uxv7/AX7qrlYD31fVjzp7HT98vmM3LxVu5/JZo8lM8dTwKjIWLvQ7AmNMFPLyFPtERP4GPIdTNAR4aj76ILAAeLid/Z8DM1V1l4icCCwEDvMQj++e+2gbAOdML/A3EGOM6QFeEkEaTgI4vtW2DpuPqupyESkIsf/frVZXAFEzNsIrhSVMGd6Pwdmpfoeyt3nznKW9GRhjOsFLz+JIzEtwEfBieztFZB4wD2DECH/L5Isra/l4ayXXnDDe1zjatGGD3xEYY6KQlw5lqTgP6onAnj+BVfXCnghARI52z39ke8eo6kKcoiOmTp2qPXHdrnq1sASA4ycM8jMMY4zpMV7aPT4CDAa+DSzDKcKp6omLu0NX3AecqqrlPXHOcHu5sISRuRmMzsv0OxRjjOkRXhLBGFW9Dtitqg8Bs4FuD6wjIiNw6hnOUdWoKNMIBBtYsamc4yYMsp7ExpiY4aWyuMFdVojIJGA7zgB0IYnIY8AsIFdEioAbgCQAVb0HuB7IAe5yH6qNqjq1k/FH1LL1ZTQ0Kcf11mKhKVP8jsAYE4W8JIKFItIfuA54Fsh0v4ekqmd3sP9i4GIvQfYWrxSWkJORzCEj+vsdStts1FFjTBd4aTV0n/t1GRC3s7I3NDXzxvpSTpg4mIQ+VixkjIkd7dYRiMjJIrJfq/XrReQjEXlWROJuYt53N+2kKtjYe4uFAL73PedjjDGdEKqy+CagDEBE5gDfAy7EKR66J/yh9S6vFG4nNakP3xzbC+YdaE9RkfMxxphOCJUIVFVr3O+nA/er6iq3qKgXPw17nqrySmEJR47JIy05we9wjDGmR4VKBCIimSLSBzgWeK3Vvl42tkJ4rd0WYFtl0DqRGWNiUqjK4vnAaiAArFPVlQAicjBQHPbIepFXCksQgWMOGOh3KMYY0+PaTQSqukhEXgIGAq2Hh94ORGL8oV7jlcISDh3Rn9zMFL9DCe3ww/2OwBgThUI2H1XVrcDWfbbF1dtA0a4aCosD/OzE/f0OpWO/+53fERhjopDNsdiBlkHmenWzUWOM6QZLBB14ZV0Jo/MyGBUNg8x997vOxxhjOqFTiUBEMkTkeyLyz3AF1JtU1jbw7qadHDdhsN+heFNe7nyMMaYTOkwEIpIsIqeJyBKc1kLHEicdypauL6WxuRcPMmeMMT2g3cpiETkOOBtnHoI3cOYlmBahGct6haXry8jNTObg4f38DsUYY8ImVKuhl4A3gSNV9XMAEflTRKLqJdZsrWTK8P70sUHmjDExLFQiOBQ4C3hVRDYBi4G4GV8h2NDEZ2XVnHjgEL9D8e7YY/2OwBgThUJ1KPsQ+BD4qYjMwCkmShaRF4Gn3HmEY9b67VU0K0wY0tfvULy7rsNpIowx5ms8tRpS1bdV9QpgGM7QEzHfhbWwOADAxKFRlAiMMaYLQlUWj2hn1zqcaSdjWuG2AFkpieT3T/M7FO9OPNFZvviiv3EYY6JKqDqCfwIKtK4pVZwhqAcS4/UFhcUBDhjaN7omqa+t9TsCY0wUardoSFUPVNXJ7vJA4GTgbaAauCpC8fmiuVlZVxyIrvoBY4zpIi8dysaKyIPAi8AqYIKq3hnuwPz0xc4aauqbLBEYY+JCqDqCScAvgInAzcBFqtoUqcD8VLjNqSieYBXFxpg4EKqO4CNgC05dwTRgWuvyclW9Mryh+aewuJLEPsKYgVEw0Fxrc+b4HYExJgqFSgQXRiyKXqZwW4AxAzNJTYqy+vCf/MTvCIwxUShUh7KH2tsnIvuFJ5zeobA4wIzRuX6HYYwxERGyslhEDheRuSIy0F2fLCJ/A96KSHQ+2FFdR0mgLjrrB2bNcj7GGNMJ7SYCEbkFWAR8F/iniNwAvAK8C4yNTHiRt87tUWwthowx8SJUHcFs4GBVDYpIf2AbMFlVP41MaP5oaTF0gCUCY0ycCFU0VKuqQQBV3QWs70wSEJFFIlIqImva2S8icoeIbBSR/4jIIZ0LPTwKiwMMzU6lf0ay36EYY0xEhHojGC0iz7ZaL2i9rqqndHDuB4EFwMPt7D8Rp4hpLHAYcLe79FXhtkB01g8YY0wXhUoEp+6zfltnTqyqy0WkoIPzP6yqCqwQkX4iMkRViztznZ4UbGhi047dnDgpSuYo3tcZZ/gdgTEmCoVqProMQEQGqmpp630iMr4Hrj0Mp8NaiyJ3m2+JYENJFU3NGr31A5df7ncExpgo5GU+gjdFZM+fmiLyY+CpHrh2W8N6apsHiswTkZUisrKsrKwHLt22qB9aoqbG+RhjTCd4SQSzgHNE5O8ishwYhzPkRHcVAcNbrefjtEz6GlVdqKpTVXVqXl5eD1y6bYXFATJTEhnePz1s1wirk05yPsYY0wkdJgK3zP5fOLOSFeCU61f3wLWfBc51Ww9NByr9rB8A543ggCFZNlm9MSauhKosBkBEXsEpt5+E81f7IhFZrqohB7YRkcdw3iZyRaQIZ1azJABVvQd4ATgJ2AjUABd0/Ta6r2UOgrmH5vsZhjHGRFyHiQD4s6o+7X6vEJHDgZ939CNVPbuD/Qr8wMP1I+LLnTXsrm+K3voBY4zpIi9FQ0+LyJEi0vIXe3/g0fCGFXmFe4aWyPY5EmOMiSwvRUM3AFOB8cADQDJOIpgR3tAiq3BbgIQ+wthBUTYHQWvnn+93BMaYKOSlaOg7wMHABwCquk1EssIalQ8KiwOMyYvCOQhas0RgjOkCL81H693yfAUQkYzwhuSPmBhaYscO52OMMZ3g5Y1giYjcC/QTkUtwZi77S3jDiqzy6jq2B4LRP/T03LnOculSX8MwxkSXDhOBqt4qIscBAZx6gutV9ZWwRxZB64qrgCjuUWyMMd3g5Y0A98EfUw//1gqLKwGbg8AYE5+81BHEvMJtAQb3TWWAzUFgjIlDlghwWgxZsZAxJl6FmrP4NXf5h8iFE3mNTc18Vrab8YNjoEXs97/vfIwxphNC1REMEZGZwCkisph9ho1W1Q/CGlmE7Kiup6lZGdYvze9Quu/MM/2OwBgThUIlguuBa3EGmrt9n30KHBOuoCKpJBAEYFDfVJ8j6QFb3Hl+hg8PfZwxxrQSaoayJ4AnROQ6Vf1NBGOKqO1uIhgcC4ngnHOcpfUjMMZ0gpd+BL8RkVOAo9xNS1X1+fCGFTmle94IUnyOxBhj/NFhqyER+R3wQ6DQ/fzQ3RYTtgeCJPQRcjItERhj4pOXDmWzgSmq2gwgIg8BHwI/C2dgkVISqCMvM4UEm5XMGBOnvPYj6Nfqe0wN2F8SCDIoOwbqB4wxpou8vBH8DvhQRN7AaUJ6FDHyNgBOIijIiZEBVX/8Y78jMMZEIS+VxY+JyFLgGziJ4Kequj3cgUVKSaCOw0bm+B1Gzzj5ZL8jMMZEIa+DzhUDz4Y5logLNjRRWdvA4FgpGlq/3lmOH+9vHMaYqOIpEcSqls5kA7NipMXQpZc6S+tHYIzphLgedG57pduZLFbeCIwxpgtCJgIR6SMiayIVTKSVVNUBMTK8hDHGdFHIROD2HfhIREZEKJ6IKo2lcYaMMaaLvNQRDAHWish7wO6Wjap6StiiipDtlUFSk/rQNzWuq0qMMXHOyxPw12GPwiclVXUM6puKSIz0Kv7lL/2OwBgThbz0I1gmIvsBY1X1VRFJBxLCH1r4lVQGY6tY6Fvf8jsCY0wU8jLo3CXAE8C97qZhwNNhjCliSqpiLBGsXu18jDGmE7w0H/0BMAMIAKjqp8DAcAYVCarK9sogg2Np+OmrrnI+xhjTCV4SQZ2q1resiEgizgxlHRKRE0RkvYhsFJFr29ifLSLPichHIrJWRC7wHnr3BGobqWtsjq03AmOM6QIviWCZiPwcSBOR44C/A8919CMRSQD+DJwITADOFpEJ+xz2A6BQVQ8CZgG3iUhyJ+LvspIqt1exJQJjTJzzkgiuBcqAj4FLgRcAL81TpgEbVXWT+0axGDh1n2MUyBKn2U4msBNo9Bh7t+zpVWyJwBgT57y0Gmp2J6N5F+fBvV5VvRQNDQO2tFovAg7b55gFOIPZbQOygDNbJsBpTUTmAfMARozomb5tJTZFpTHGAB4SgYjMBu4BPsMZhnqkiFyqqi929NM2tu2bQL4NrAaOAUYDr4jIm6oa2OtHqguBhQBTp071VD/RkZJY7FX829/6HYExJgp56VB2G3C0qm4EEJHRwD+BjhJBETC81Xo+zl/+rV0A/N59w9goIp8D+wPveYirW0oCdWSnJZGaFBNdIhxHHOF3BMaYKOSljqC0JQm4NgGlHn73PjBWREa6FcBn8fU5Db4EjgUQkUHAePf8YVcSCMZe/cC//+18jDGmE9p9IxCR092va0XkBWAJTtHOf+E85ENS1UYRuQJ4Cacn8iJVXSsil7n77wF+AzwoIh/z1exnO7pzQ16VBIIMjLX6gZ//3FnafATGmE4IVTTUet7DEmCm+70M6O/l5Kr6Ak4ro9bb7mn1fRtwvKdIe1hJoI6xg7L8uLQxxvQq7SYCVY1Y565Ia2pWyqrrYq9oyBhjusBLq6GRwP8CBa2Pj+ZhqMur62hqVms6aowxeGs19DRwP05v4q+18Y9G22Ox6agxxnSRl0QQVNU7wh5JBJUEYnSKyvnz/Y7AGBOFvCSCP4nIDcDLQF3LRlX9IGxRhVlLZ7KYm7R+yhS/IzDGRCEvieBA4Byc3r8tRUPqrkelkkCQPgI5GREZ3y5yXn3VWdoENSYKNDQ0UFRURDAY9DuUmJKamkp+fj5JSUmef+MlEXwHGNV6KOpoVxIIkpuZQmKCl/50UeTGG52lJQITBYqKisjKyqKgoCB2pov1mapSXl5OUVERI0eO9Pw7L0/Cj4B+XQ2sN9oeqIu9YiFjokwwGCQnJ8eSQA8SEXJycjr9luXljWAQ8ImIvM/edQRR23y0NBAkv3+632EYE/csCfS8rvw39ZIIbuh8KL1bSSDI1AJPnaONMSbmeZmPYFkkAomUYEMTu2oaGJRlRUPGmPCaP38+8+bNIz29d5dAdFhHICJVIhJwP0ERaRKRQEe/663Kqtw+BLFYR3Dvvc7HGNMrzJ8/n5qaGr/D6JCXN4K9RmYTkdNwpqGMSjHdq3j8eL8jMKZLfv3cWgq39ezflxOG9uWGkyd2eNzDDz/MrbfeiogwefJkbrzxRi688ELKysrIy8vjgQceYMSIEZx//vnMmTOHuXPnApCZmUl1dTVLly7lV7/6Fbm5uaxZs4ZDDz2URx99lDvvvJNt27Zx9NFHk5ubyxtvvNGj99eTvNQR7EVVnxaRa8MRTCTE9BSVzz3nLE8+OfRxxhgA1q5dy0033cTbb79Nbm4uO3fu5LzzzuPcc8/lvPPOY9GiRVx55ZU8/fTTIc/z4YcfsnbtWoYOHcqMGTN4++23ufLKK7n99tt54403yM3NjcwNdZGXQedOb7XaB5jK16ecjBoxPWn9bbc5S0sEJsp4+cs9HF5//XXmzp2750E9YMAA3nnnHZ588kkAzjnnHK655poOzzNt2jTy8/MBmDJlCps3b+bII48MX+A9zMsbQeunSiOwGTg1LNFEQGlVHcmJfchO897rzhgTm1S1w+aWLfsTExNpbm7e87v6+q/62KakfFXCkJCQQGNjYxiiDZ8OK4tV9YJWn0tU9SZV9TJVZa/UMkWltV82xhx77LEsWbKE8vJyAHbu3MkRRxzB4sWLAfjrX/+65y/7goICVq1aBcAzzzxDQ0NDh+fPysqiqqoqTNH3nFBTVV4f4neqqr8JQzxht70yGJv1A8aYTps4cSK/+MUvmDlzJgkJCRx88MHccccdXHjhhdxyyy17KosBLrnkEk499VSmTZvGscceS0ZGRofnnzdvHieeeCJDhgzp1ZXFotp2cb+I/LiNzRnARUCOqmaGM7D2TJ06VVeuXNnl3x9961ImDu3Lgv8+pAej6iVmzXKWNmexiQLr1q3jgAMO8DuMmNTWf1sRWaWqU9s6PtRUlbe1OkEW8EPgAmAxcFt7v+vNVJXtlUGO2X+g36GExyOP+B2BMSYKhawsFpEBwI+A/wEeAg5R1V2RCCwcquoaqW1oit2ioeHD/Y7AGBOFQtUR3AKcDiwEDlTV6ohFFSalsdyZDODxx53lmWf6G4cxJqqEeiP4Mc5oo78EftGqlY3gVBb3DXNsPW57ZYxOUdni7rudpSUCY0wnhKojiLFZW1pNURmricAYY7og5h72ocT0OEPGGNNFcZUISgNB+qYmkpac4HcoxhifVVRUcNddd+1ZX7p0KXPmzOnx65x//vk88cQTnfrNggULGDNmDCLCjh079mxXVa688krGjBnD5MmT+eCDD3okxrhKBNsDQXsbMMYAX08EXjU1NYUhmr3NmDGDV199lf3222+v7S+++CKffvopn376KQsXLuT73/9+j1yv06OPRrOSQF1sJ4JO/tVhTK/S0iGytTPOgMsvh5oaOOmkr+8//3zns2MHuMND79FBx8prr72Wzz77jClTpnDccccxe/ZsqqurmTt37l7DSYsIBQUFXHjhhbz88stcccUVDBgwgBtuuIG6ujpGjx7NAw88QGZmJtdeey3PPvssiYmJHH/88dx6660ALF++nNtvv53t27dz88037xnKuj0HH3xwm9ufeeYZzj33XESE6dOnU1FRQXFxMUOGDAl5vo6ENRGIyAnAn4AE4D5V/X0bx8wC5gNJwA5VnRmueEoDQUaP7t3DwXZLLx/q1pje5Pe//z1r1qxh9erVgFM01NZw0i1jDaWmpvLWW2+xY8cOTj/9dF599VUyMjL4wx/+wO23384VV1zBU089xSeffIKIUFFRsedaxcXFvPXWW3zyySeccsopexLBlClT9lzfi61btzK8VX+h/Px8tm7d2nsTgYgkAH8GjgOKgPdF5FlVLWx1TD/gLuAEVf1SRMLW5be5WSmtqovdzmQADz7oLM8/388ojOmaUH/Bp6eH3p+b2yNDq4QaTvpMt1n2ihUrKCwsZMaMGQDU19dz+OGH07dvX1JTU7n44ouZPXv2XvUNp512Gn369GHChAmUlJTs2d6ZJABOHcG+emIAzXC+EUwDNqrqJgARWYwzfHVhq2P+G3hSVb8ECOeopuW762lsVgbH4hSVLSwRGNMtoYaTbhlkTlU57rjjeOyxx772+/fee4/XXnuNxYsXs2DBAl5//fWvnbe98d28yM/PZ8uWLXvWi4qKGDp0aJfP1yKclcXDgC2t1ovcba2NA/qLyFIRWSUi57Z1IhGZJyIrRWRlWVlZl4Jp6UMw0CatN8bQ9SGip0+fzttvv83GjRsBqKmpYcOGDVRXV1NZWclJJ53E/PnzO/3XvhennHIKDz/8MKrKihUryM7O7naxEIQ3EbT1vrJvKkwEDgVmA98GrhORcV/7kepCVZ2qqlPz8vK6FMyezmSx/EZgjPEsJyeHGTNmMGnSJK6++mrPv8vLy+PBBx/k7LPPZvLkyUyfPp1PPvmEqqoq5syZw+TJk5k5cyZ//OMfOzzXlClT2tx+xx13kJ+fT1FREZMnT+biiy8G4KSTTmLUqFGMGTOGSy65pEutntrS7jDU3T6xyOHAr1T12+76zwBU9XetjrkWSFXVX7nr9wP/UtW/t3ferg5DvXLzTu5783P+77SJsftWYMNQmyhiw1CHT48NQ90D3gfGishIYCtwFk6dQGvPAAtEJBFIBg4DOk6jXTC1YABTCwaE49TGGBPVwpYIVLVRRK4AXsJpPrpIVdeKyGXu/ntUdZ2I/Av4D9CM08R0TbhiinkvvOB3BMaYKBTWfgSq+gLwwj7b7tln/RbglnDGETfS0/2OwJhO8TJ5vOmcrhT3x9UQEzHvrrucjzFRIDU1lfLy8m41pzR7U1XKy8tJTe1cPWhcDTER85YscZaXX+5vHMZ40NIqpqtNwk3bUlNT93SK88oSgTHGF0lJSYwcOdLvMAxWNGSMMXHPEoExxsQ5SwTGGBPnwtazOFxEpAz4ooPDcoEdHRwTq+L53iG+7z+e7x3i+/693Pt+qtrmGD1Rlwi8EJGV7XWljnXxfO8Q3/cfz/cO8X3/3b13Kxoyxpg4Z4nAGGPiXKwmgoV+B+CjeL53iO/7j+d7h/i+/27de0zWERhjjPEuVt8IjDHGeGSJwBhj4lzMJQIROUFE1ovIRncGtJglIotEpFRE1rTaNkBEXhGRT91lfz9jDBcRGS4ib4jIOhFZKyI/dLfHy/2nish7IvKRe/+/drfHxf0DiEiCiHwoIs+76/F075tF5GMRWS0iK91tXb7/mEoEIpIA/Bk4EZgAnC0iE/yNKqweBE7YZ9u1wGuqOhZ4zV2PRY3Aj1X1AGA68AP3/3W83H8dcIyqHgRMAU4QkenEz/0D/BBY12o9nu4d4GhVndKq/0CX7z+mEgEwDdioqptUtR5YDJzqc0xho6rLgZ37bD4VeMj9/hBwWiRjihRVLVbVD9zvVTgPhGHEz/2rqla7q0nuR4mT+xeRfGA2cF+rzXFx7yF0+f5jLREMA7a0Wi9yt8WTQapaDM7DEhjoczxhJyIFwMHAu8TR/btFI6uBUuAVVY2n+58PXIMzxW2LeLl3cJL+yyKySkTmudu6fP+xNh9BW3PeWfvYGCYimcA/gKtUNRBP0x6qahMwRUT6AU+JyCSfQ4oIEZkDlKrqKhGZ5XM4fpmhqttEZCDwioh80p2TxdobQREwvNV6PrDNp1j8UiIiQwDcZanP8YSNiCThJIG/quqT7ua4uf8WqloBLMWpL4qH+58BnCIim3GKf48RkUeJj3sHQFW3uctS4CmcYvEu33+sJYL3gbEiMlJEkoGzgGd9jinSngXOc7+fBzzjYyxhI86f/vcD61T19la74uX+89w3AUQkDfgW8AlxcP+q+jNVzVfVApx/46+r6veIg3sHEJEMEclq+Q4cD6yhG/cfcz2LReQknPLDBGCRqt7kb0ThIyKPAbNwhqAtAW4AngaWACOAL4H/UtV9K5SjnogcCbwJfMxX5cQ/x6kniIf7n4xTIZiA8wfdElX9PxHJIQ7uv4VbNPQTVZ0TL/cuIqNw3gLAKd7/m6re1J37j7lEYIwxpnNirWjIGGNMJ1kiMMaYOGeJwBhj4pwlAmOMiXOWCIwxJs5ZIjBxQ0Ry3NEaV4vIdhHZ6n6vEJHCMFzvVyLyk07+prqd7Q+KyNyeicyYvVkiMHFDVcvd0RqnAPcAf3S/T2HvMWvaJCKxNiSLMYAlAmNaJIjIX9yx/V92e+siIktF5Lcisgz4oYgcKiLL3MG+XmrVpf9KESkUkf+IyOJW553gnmOTiFzZslFEfiQia9zPVfsGI44F7jn/SWwPoGZ8Zn/hGOMYC5ytqpeIyBLgu8Cj7r5+qjrTHdtoGXCqqpaJyJnATcCFOGO/j1TVupahH1z7A0cDWcB6EbkbmAxcAByGM1DiuyKyTFU/bPW77wDjgQOBQUAhsCgcN26MJQJjHJ+r6mr3+yqgoNW+x93leGASzmiP4AzvUOzu+w/wVxF5GmeYjxb/VNU6oE5ESnEe6kcCT6nqbgAReRL4JtA6ERwFPOaOMLpNRF7v/i0a0zZLBMY46lp9bwLSWq3vdpcCrFXVw9v4/Wych/cpwHUiMrGd8ybS9nDpbbHxX0xEWB2BMd6tB/JE5HBwhsEWkYki0gcYrqpv4EyW0g/IDHGe5cBpIpLujh75HZwB9PY95ix38pkhOMVLxoSFvREY45Gq1rtNOO8QkWycfz/zgQ3Ao+42wWmNVNHeJDmq+oGIPAi85266b5/6AXBGlzwGZ3TVDTh1E8aEhY0+aowxcc6KhowxJs5ZIjDGmDhnicAYY+KcJQJjjIlzlgiMMSbOWSIwxpg4Z4nAGGPi3P8HeGbnuDW4EmEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot('threshold', 'count', data=exRNA_filter_results)\n",
    "plt.xlabel(\"Threshold\")\n",
    "plt.ylabel(\"Number of exRNA Samples Overlap\")\n",
    "ax = plt.gca()\n",
    "cutoff=10\n",
    "ax.axvline(cutoff, color = \"r\", ls=\"--\", label=f\"thresh: {cutoff}\")\n",
    "plt.legend()\n",
    "plt.savefig(\"figures/exRNA_atlas_samples_filter_thresh.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we choose a cutoff of 10. If a locus is found (overlapped) in at least 10 exRNA samples, we will remove from consideration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "275205"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_remove = set(exRNA_loci_hits_df[exRNA_loci_hits_df[\"counts\"] >= 10][\"loci\"])\n",
    "len(to_remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1459094"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_filtered_loci = []\n",
    "for locus in loci_map:\n",
    "    if locus not in to_remove:\n",
    "        exRNA_filtered_loci.append(locus)\n",
    "len(exRNA_filtered_loci)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this exRNA cleanup step, we removed about 15% of our feature loci. Removing 275205 loci from 1734299."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed\", \"w\") as out: \n",
    "    for locus in exRNA_filtered_loci:\n",
    "        splits = locus.split(\":\")\n",
    "        start = splits[1].split(\"-\")[0]\n",
    "        end = splits[1].split(\"-\")[1]\n",
    "        bed = f\"{splits[0]}\\t{start}\\t{end}\\t{locus}\\t.\\t{splits[-1]}\"\n",
    "        out.write(bed + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Counts\n",
    "Last preprocess step for the cell lines is to create counts for each cell line sample for each smRNA loci feature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for f in data/IC/*.filter.bed;\n",
    "do\n",
    "out=${f/.filter.bed/.exRNA.intersect.bed}\n",
    "echo \"intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed > $out\"\n",
    "intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed > $out\n",
    "done &> log/cell_lines_exRNA_intersect.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for f in data/IC/*.filter.bed;\n",
    "do\n",
    "out=${f/.filter.bed/.intersect.bed}\n",
    "echo \"intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map.bed > $out\"\n",
    "intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map.bed > $out\n",
    "done &> log/cell_lines_intersect.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_loci = {}\n",
    "cpm_map = {}\n",
    "exRNA_filtered_cell_lines = [f for f in os.scandir(\"data/IC/\") if f.name.endswith(\".exRNA.intersect.bed\")]\n",
    "for f in exRNA_filtered_cell_lines:  \n",
    "    \n",
    "    sample = f.name.split(\".\")[0]\n",
    "    bed_file = pd.read_csv(f\"data/IC/{sample}.filter.bed\", header=None, sep=\"\\t\")\n",
    "    assert len(bed_file[3].unique()) == bed_file.shape[0] #Ensures number of unique reads (query_id) matches total number of reads in bed file, so that we are not over counting\n",
    "    #cpm calculations: total reads within pre-exRNA filtered bedfile. Note reads that had \"chrUn\" or low-complexity seqeunces where filered out and not included in cpm normalization.           \n",
    "    cpm = 1000000/bed_file.shape[0]\n",
    "    cpm_map[sample] = cpm\n",
    "    \n",
    "    #Counts of exRNA filtered loci\n",
    "    ex_fil_bed = pd.read_csv(f, header=None, sep=\"\\t\")\n",
    "    loci_features = ex_fil_bed[9] #Locus annotation as a feature.\n",
    "    loci_bp_overlaps = ex_fil_bed[12] #Number of bp overlaps between read and locus annotation.\n",
    "    read_ids = ex_fil_bed[3] #Query ID from original bamfile for each read.\n",
    "    \n",
    "    #First create read_id map\n",
    "    read_id_locus_map = {}\n",
    "    for i in range(len(read_ids)):\n",
    "        _id = read_ids.iloc[i]\n",
    "        num_bp_match = loci_bp_overlaps.iloc[i]\n",
    "        locus = loci_features.iloc[i]\n",
    "        \n",
    "        if _id in read_id_locus_map: #Indicates multiple hits/overlaps for one read.\n",
    "            if num_bp_match > read_id_locus_map[_id][\"bp\"]: #Update locus count if more number of bp matched. This is our simple binning procedure.\n",
    "                read_id_locus_map[_id] = {\"locus\":locus, \"bp\":num_bp_match}\n",
    "        else:\n",
    "            read_id_locus_map[_id] = {\"locus\":locus, \"bp\":num_bp_match}\n",
    "            \n",
    "    assert len(read_id_locus_map) == len(ex_fil_bed[3].unique()) #Ensures we do not overcount reads   \n",
    "    \n",
    "    #Counts of loci\n",
    "    sample_loci[sample] = {}\n",
    "    for _id in read_id_locus_map: \n",
    "        locus = read_id_locus_map[_id][\"locus\"] \n",
    "        if locus in sample_loci[sample]:\n",
    "            sample_loci[sample][locus] += 1\n",
    "        else:\n",
    "            sample_loci[sample][locus] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/counts/cell_lines_loci_counts.json', 'w') as f:\n",
    "    js.dump(sample_loci, f)\n",
    "    f.close()\n",
    "    \n",
    "with open (\"data/counts/cell_lines_cpm_map.json\", \"w\") as f:\n",
    "    js.dump(cpm_map, f)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess TCGA Data\n",
    "Since TCGA datasets are quite bit larger, we run the following pre-processing steps as scripts instead of in this notebook. All scripts can be found in the `scripts` directory. <br>\n",
    "TCGA dataset we use in this project involves: 679 normal samples from varying tissues and 1103 BRCA cancer samples.\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dust \n",
    "Eliminate low-complexity sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`python3 scripts/dust_filter.py &> log/dust_filter_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bamtobed\n",
    "Convert all the filtered bam files to bed files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`ls /rumi/shams/jwang/BRCA_oncRNA/data/TCGA/*.dust.bam | parallel -j 30 bash scripts/bamtobed.sh {} &> log/bamtobed_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter TCGA Data\n",
    "TCGA bed data contains certain formats that are not compatible with bedtools and are not of interest (ex: chrUn). Here we filter our TCGA data before further processing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`python3 scripts/filter_TCGA.py &> log/filter_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "Finished preprocessing and creating counts for loci found in our cell lines and preprocessing TCGA data. <br>\n",
    "Next step is identify cancer-enriched RNAs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "# Below Code Needs to Be Refined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "### Merge TCGA BRCA Loci: Feature Annotation\n",
    "Run in terminal in smRNA env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "tcga_brca_loci = set()\n",
    "TCGA_filtered = [f for f in os.scandir(\"data/TCGA/\") if f.name.endswith(\"filter.bed\")]\n",
    "for f in TCGA_filtered:\n",
    "    with open(f, \"r\") as filtered_bed:\n",
    "        for line in filtered_bed:\n",
    "            splits = line.split(\"\\t\")\n",
    "            bed_locus = f\"{splits[0]}\\t{splits[1]}\\t{splits[2]}\\t.\\t.\\t{splits[5].strip()}\"\n",
    "            tcga_brca_loci.add(bed_locus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "brca_loci = pd.DataFrame(list(tcga_brca_loci))\n",
    "brca_loci.to_csv(\"data/TCGA_BRCA_all_loci.bed\", header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6\\t28543111\\t28543134\\t.\\t.\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5\\t119541975\\t119542002\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr15\\t28215223\\t28215248\\t.\\t.\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1\\t151994373\\t151994400\\t.\\t.\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20\\t60316075\\t60316094\\t.\\t.\\t-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82558337</th>\n",
       "      <td>chr1\\t93346106\\t93346126\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82558338</th>\n",
       "      <td>chr18\\t3450234\\t3450252\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82558339</th>\n",
       "      <td>chr6\\t161065280\\t161065298\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82558340</th>\n",
       "      <td>chr9\\t129635173\\t129635193\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82558341</th>\n",
       "      <td>chr4\\t123870378\\t123870398\\t.\\t.\\t+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82558342 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            0\n",
       "0           chr6\\t28543111\\t28543134\\t.\\t.\\t-\n",
       "1         chr5\\t119541975\\t119542002\\t.\\t.\\t+\n",
       "2          chr15\\t28215223\\t28215248\\t.\\t.\\t-\n",
       "3         chr1\\t151994373\\t151994400\\t.\\t.\\t-\n",
       "4          chr20\\t60316075\\t60316094\\t.\\t.\\t-\n",
       "...                                       ...\n",
       "82558337    chr1\\t93346106\\t93346126\\t.\\t.\\t+\n",
       "82558338     chr18\\t3450234\\t3450252\\t.\\t.\\t+\n",
       "82558339  chr6\\t161065280\\t161065298\\t.\\t.\\t+\n",
       "82558340  chr9\\t129635173\\t129635193\\t.\\t.\\t+\n",
       "82558341  chr4\\t123870378\\t123870398\\t.\\t.\\t+\n",
       "\n",
       "[82558342 rows x 1 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brca_loci"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "We have 82558342 unique loci in our TCGA BRCA dataset. Next step is to merge to create our TCGA BRCA loci feature annotation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "`cat data/TCGA_BRCA_loci.bed | sort -k1,1 -k2,2n -T temp | mergeBed -s -c 6 -o distinct -i stdin | \\\n",
    "awk '{print $1 \"\\t\" $2 \"\\t\" $3 \"\\t\" $1\":\"$2\"-\"$3\":\"$4  \"\\t.\" \"\\t\" $4  > \"data/TCGA_BRCA_all_smRNAs_merged_loci.bed\"}' `"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:smRNA]",
   "language": "python",
   "name": "conda-env-smRNA-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
