{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Breast Cancer Cell Lines: Preprocess Data\n",
    "In this notebook we preprocess our data for downstream locus-based analysis. Data includes:\n",
    "* smRNA seq data of the breast cancer (and normal) cell lines used in our Nature Med paper (Fish, 2018)\n",
    "* smRNA seq data from TCGA: specifically all the BRCA samples and all the normal samples from all tissue types in TCGA. \n",
    "* smRNA seq data from exRNA Atlas along with other publically available datasets from non-cancerous exosomal smRNA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pysam\n",
    "import re\n",
    "import json as js\n",
    "from scipy.signal import find_peaks, peak_widths\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Cell Lines Data\n",
    "Dataset involves 21 samples:\n",
    "* 3 samples from HMEC cell lines\n",
    "* 18 samples from "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_lines = [f for f in os.scandir(\"data/IC/\") if f.name.endswith(\".srt.dd.bam\")]\n",
    "len(cell_lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dust"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we are converting the bam files to bed format, we first remove the low complexity sequences in the bam files first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Simple dust score adapted from Dust Algorithm Score: A fast and symmetric DUST implementation to mask\n",
    "low-complexity DNA sequences (2006, Morgulis et al.)\"\"\"\n",
    "def simpleDustScore(seq):\n",
    "    assert len(seq) > 2\n",
    "    if len(seq) == 3:\n",
    "        return 0\n",
    "    else:\n",
    "        triplets = {}\n",
    "        num_trip = len(seq) - 2\n",
    "        for i in range(num_trip):\n",
    "            subseq = seq[i:i+3]\n",
    "            if subseq in triplets:\n",
    "                triplets[subseq] += 1\n",
    "            else:\n",
    "                triplets[subseq] = 1\n",
    "        sum_triplet = 0\n",
    "        for triplet, count in triplets.items():\n",
    "            sum_triplet += count * (count - 1) / 2\n",
    "        return sum_triplet/(num_trip - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in cell_lines: \n",
    "    name = f.name.split(\".\")[0]\n",
    "    out = f\"data/IC/{name}.srt.dd.dust.bam\"\n",
    "    infile = pysam.AlignmentFile(f, \"rb\")\n",
    "    outfile = pysam.AlignmentFile(out, \"wb\", template=infile)\n",
    "    for read in infile.fetch():\n",
    "        if simpleDustScore(read.get_forward_sequence()) < 3 and len(read.get_forward_sequence()) >= 15:\n",
    "            outfile.write(read)\n",
    "    outfile.close()\n",
    "    infile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BedtoBam\n",
    "Next we convert all the cell lines bam alignment files to bedfiles. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "for f in data/IC/*.srt.dd.dust.bam;\n",
    "do \n",
    "base=$(basename $f)\n",
    "out=${base/.bam/.bed}\n",
    "echo \"$bedtools bamtobed -i $f > data/IC/$out\"\n",
    "bedtools bamtobed -i $f > data/IC/$out\n",
    "done &> log/cell_lines_bamtobed.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter Cell Lines Bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in os.scandir(\"data/IC\"):\n",
    "    if f.name.endswith(\".srt.dd.dust.bed\"):\n",
    "        name = f.name.split(\".\")[0]\n",
    "        outfile = f\"data/IC/{name}.filter.bed\"\n",
    "        with open(outfile, \"wt\") as out, open(f, \"rt\") as file:\n",
    "            for line in file:\n",
    "                if re.match(\"chr[\\d+,X,Y]\", line) and \"None\" not in line:\n",
    "                    out.write(line) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge Cell Lines Loci\n",
    "Here we merge all the loci seen in the cell lines data to create a locus feature map. No threshold set for degree of overlap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat data/IC/*.filter.bed | sort -k1,1 -k2,2n | mergeBed -s -c 6 -o distinct -i stdin | \\\n",
    "awk '{print $1 \"\\t\" $2 \"\\t\" $3 \"\\t\" $1\":\"$2\"-\"$3\":\"$4  \"\\t.\" \"\\t\" $4  > \"data/cell_lines_smRNAs_merged_loci.bed\"}' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split Loci By Peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since some loci are larger than 200 bp, we will need to split these loci up by coverage. <br>\n",
    "First we merge all the cell lines sample bamfiles to run `bedtools genomecov`. Based on coverage peaks we will separate the larger loci into smaller ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "samtools merge data/IC/all_samples.bam data/IC/*srt.dd.dust.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools genomecov -ibam data/IC/all_samples.bam -bg -strand - > data/IC/all_samples_neg_bg.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools genomecov -ibam data/IC/all_samples.bam -bg -strand + > data/IC/all_samples_pos_bg.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629083</td>\n",
       "      <td>630075</td>\n",
       "      <td>chr1:629083-630075:+</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629218</td>\n",
       "      <td>629905</td>\n",
       "      <td>chr1:629218-629905:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630084</td>\n",
       "      <td>630520</td>\n",
       "      <td>chr1:630084-630520:+</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630317</td>\n",
       "      <td>630520</td>\n",
       "      <td>chr1:630317-630520:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630645</td>\n",
       "      <td>631429</td>\n",
       "      <td>chr1:630645-631429:-</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0       1       2                     3  4  5  length\n",
       "224  chr1  629083  630075  chr1:629083-630075:+  .  +     992\n",
       "225  chr1  629218  629905  chr1:629218-629905:-  .  -     687\n",
       "228  chr1  630084  630520  chr1:630084-630520:+  .  +     436\n",
       "231  chr1  630317  630520  chr1:630317-630520:-  .  -     203\n",
       "237  chr1  630645  631429  chr1:630645-631429:-  .  -     784"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_loci = pd.read_csv(\"data/cell_lines_smRNAs_merged_loci.bed\", header=None, sep=\"\\t\")\n",
    "all_loci[\"length\"] = all_loci[2] - all_loci[1] #Calculate length of each locus\n",
    "loci_to_split = all_loci[all_loci[\"length\"] > 200]\n",
    "loci_to_split.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11046</td>\n",
       "      <td>11061</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11237</td>\n",
       "      <td>11258</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11768</td>\n",
       "      <td>11785</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12036</td>\n",
       "      <td>12072</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17017</td>\n",
       "      <td>17035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2  3\n",
       "0  chr1  11046  11061  1\n",
       "1  chr1  11237  11258  1\n",
       "2  chr1  11768  11785  1\n",
       "3  chr1  12036  12072  1\n",
       "4  chr1  17017  17035  1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_density = pd.read_csv(\"data/IC/all_samples_pos_bg.txt\", header=None, sep=\"\\t\")\n",
    "neg_density = pd.read_csv(\"data/IC/all_samples_neg_bg.txt\", header=None, sep=\"\\t\")\n",
    "pos_density.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4400, 1011)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_loci_list = []\n",
    "for i in range(loci_to_split.shape[0]):\n",
    "    rna = loci_to_split.iloc[i]\n",
    "    density = neg_density if rna[5] == \"-\" else pos_density\n",
    "    density = density[density[0] == rna[0]] #Match reference chr\n",
    "    density = density[(density[1] >= rna[1]) & (density[2] <= rna[2])] #Match rna locus. Since the loci area already merged, the coverage start and end positions should match seamlessly.\n",
    "    coord, coverage = [], []\n",
    "    assert density.iloc[0][1] == rna[1]\n",
    "    assert density.iloc[-1][2] == rna[2] \n",
    "    posit = density.iloc[0][1] #Start of locus\n",
    "    for i in range(density.shape[0]):\n",
    "        row = density.iloc[i]\n",
    "        start, end, count = row[1], row[2], row[3]\n",
    "        \n",
    "        #Mergebed and genomecov results have about 6 loci that are off by 1~3 bps. Here we fix continuity  \n",
    "        while posit < start:\n",
    "            coord.append(posit)\n",
    "            coverage.append(0) #Bridge gap in genomecov with 0s\n",
    "            posit += 1\n",
    "            \n",
    "        #Generating coverage by each nucletotide position.\n",
    "        while start < end:\n",
    "            coord.append(start)\n",
    "            coverage.append(count)\n",
    "            start += 1\n",
    "            posit += 1\n",
    "            \n",
    "    peaks, _ = find_peaks(np.array(coverage), height=5, rel_height=0.5, width=[15, 200], distance=20)\n",
    "    results = peak_widths(np.array(coverage), peaks, rel_height=0.5)\n",
    "    #peak_widths returns 4 arrays in order: peak widths, peak heights, peak start, peak end.\n",
    "    for new_start_index,new_end_index in zip(results[2], results[3]):\n",
    "        new_start_index, new_end_index = int(new_start_index), int(new_end_index) #These are the new indices\n",
    "        new_start = coord[new_start_index]\n",
    "        new_end = coord[new_end_index]\n",
    "        \n",
    "        if new_end - new_start + 7 < 200: #Add an additional 6 nt buffer if loci length is not too large.\n",
    "            new_start -= 3\n",
    "            new_end += 4 #1 nt more for end because bed format is end exclusive.\n",
    "        new_locus = f\"{rna[0]}:{new_start}-{new_end}:{rna[5]}\"\n",
    "        new_loci_list.append(new_locus)\n",
    "        \n",
    "len(new_loci_list), loci_to_split.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the 1011 loci greater than 200 bp, we split them into 4400 loci."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1734531"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_list = all_loci[all_loci[\"length\"] <= 200][3].tolist()\n",
    "loci_list.extend(new_loci_list)\n",
    "len(loci_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/cell_lines_smRNAs_split_loci.bed\", \"wt\") as out:   \n",
    "    for locus in loci_list:\n",
    "        splits = locus.split(\":\")\n",
    "        start = splits[1].split(\"-\")[0]\n",
    "        end = splits[1].split(\"-\")[1]\n",
    "        bed = f\"{splits[0]}\\t{start}\\t{end}\\t{locus}\\t.\\t{splits[-1]}\"\n",
    "        out.write(bed + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat data/cell_lines_smRNAs_split_loci.bed | sort -k1,1 -k2,2n | mergeBed -s -c 6 -o distinct -i stdin | \\\n",
    "awk '{print $1 \"\\t\" $2 \"\\t\" $3 \"\\t\" $1\":\"$2\"-\"$3\":\"$4  \"\\t.\" \"\\t\" $4  > \"data/cell_lines_smRNAs_loci_map.bed\"}' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1734299"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_map = pd.read_csv(\"data/cell_lines_smRNAs_loci_map.bed\", header=None, sep=\"\\t\")[3]\n",
    "len(loci_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Final number of loci in our feature map is 1734299."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## exRNA Filter Feature Map\n",
    "Here we filter out all the loci that are also seen in non-cancerous samples from the exRNA Atlas and other publically available non-cancerous exosomal smRNAseq datasets. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command: <br>\n",
    "`ls /rumi/shams/jwang/exRNA_atlas/data/bedfiles/*.srt.filter.bed | parallel -j 30 bash scripts/intersect_exRNA_atlas_loci.sh {} &> log/intersect_exRNA_atlas_loci.out` <br>\n",
    "\n",
    "\n",
    "The main command in the script is: <br>\n",
    "`intersectBed -s -u -a data/cell_lines_smRNAs_loci_map.bed -b $1 > data/exRNA/$out` <br>\n",
    "The `-u` option reports a locus in `-a` **once** when there is at least one overlap with `-b`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot exRNA clean up threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1383"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_intersect_files = [f for f in os.scandir(\"data/exRNA\") if os.stat(f).st_size != 0 and f.name.endswith(\".bed\")] #Exclude empty no hit samples.\n",
    "len(exRNA_intersect_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1143299"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_loci_hits = {}\n",
    "for file in exRNA_intersect_files:\n",
    "    df = pd.read_csv(file, sep=\"\\t\", header=None)\n",
    "    for locus in df[3]:\n",
    "        if locus in exRNA_loci_hits:\n",
    "            exRNA_loci_hits[locus] += 1\n",
    "        else:\n",
    "            exRNA_loci_hits[locus] = 1\n",
    "len(exRNA_loci_hits) #Represents number of loci in our loci feature map that had at least one overlap with a sample in the exRNA atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>loci</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:629497-629544:+</td>\n",
       "      <td>1098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:629535-629573:-</td>\n",
       "      <td>1092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:629567-629614:+</td>\n",
       "      <td>1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:630746-630788:+</td>\n",
       "      <td>1027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:630755-630794:-</td>\n",
       "      <td>799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   loci  counts\n",
       "0  chr1:629497-629544:+    1098\n",
       "1  chr1:629535-629573:-    1092\n",
       "2  chr1:629567-629614:+    1187\n",
       "3  chr1:630746-630788:+    1027\n",
       "4  chr1:630755-630794:-     799"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_loci_hits_df = pd.DataFrame.from_dict(exRNA_loci_hits, orient=\"index\", columns=[\"counts\"])\n",
    "exRNA_loci_hits_df = exRNA_loci_hits_df.reset_index(drop=False).rename(columns={\"index\":\"loci\"})\n",
    "exRNA_loci_hits_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_loci_num = 1734299\n",
    "exRNA_filter_results = {\"threshold\":[], \"count\":[]}\n",
    "for th in range(1, 50):\n",
    "    exRNA_filter_results[\"threshold\"].append(th)\n",
    "    exRNA_filter_results[\"count\"].append(total_loci_num - np.sum(exRNA_loci_hits_df[\"counts\"] >= th)) #Number of total loci minus number of loci eliminated (loci that intersected with or more than th samples) given a threshold TH\n",
    "exRNA_filter_results = pd.DataFrame(exRNA_filter_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAERCAYAAAB2CKBkAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAyAklEQVR4nO3deXhU9fX48fch+w4kYQ0YdgVE1BRFXHAtuNdSl1rXVrT2+7V+61LbWq2ttv3VaqnFjSrFpYJo3eu+AJWKCogKQRARJSQkIUAmIZms5/fHncQAyeQGMnMzM+f1PPe5c5e5c64+3JP7WUVVMcYYE7t6eR2AMcYYb1kiMMaYGGeJwBhjYpwlAmOMiXGWCIwxJsZZIjDGmBgXkYlAROaKSJmIrHZ5/rkiUigia0TkiVDHZ4wxkUQisR+BiBwLVAOPqur4Ts4dBSwETlDVHSLST1XLwhGnMcZEgoh8I1DVJcD2tvtEZISIvCoiK0TkPyJyYODQFcC9qroj8F1LAsYY00ZEJoIOzAH+V1UPB64H7gvsHw2MFpGlIrJMRKZ5FqExxvRA8V4H0B1EJB04CnhKRFp2JwXW8cAoYCqQB/xHRMar6s4wh2mMMT1SVCQCnDebnao6sZ1jRcAyVW0AvhSRdTiJ4cMwxmeMMT1WVBQNqaoP5yH/PQBxHBI4/BxwfGB/Dk5R0UYv4jTGmJ4oIhOBiMwH3gPGiEiRiPwQuBD4oYh8DKwBzgqc/hpQISKFwDvADapa4UXcxhjTE0Vk81FjjDHdJyLfCIwxxnSfiKsszsnJ0fz8fK/DMD3RunXOeswYb+MwpgdasWLFNlXNbe9YxCWC/Px8li9f7nUYpieaOtVZL1rkZRTG9Egi8lVHx6xoyBhjYlzEvREY06Gbb/Y6AmMikiUCEz1OOsnrCIyJSFY0ZKLHqlXOYozpEnsjMNHj2mudtVUWG9Ml9kZgjDExzhKBMcbEOCsaMsYYDzQ3K/7GJmrqm6itb6K2wflcU9+Iv+Gb/a2fG5o4/IA+HDOq3T5h+8USgTHGdEBVqW1oYled84BuWbc8sHfVNVHT0ERNXWPrw7rleG39ng/5RvwNzbt97qofTx1hicCYoH7/e68jMD1AY1Mz1XWNVPkbqa4LLP5Gquoa2dXmc7W/kZp653hNfRPVgeM19U3frOsb6cq4nMkJvUhNjCclIY60pDhSEuNJTYgjNyOJlMRUUhLiSEmIIzUxjuTA+pvP8bvtT0l0zk1pOSc+jl69pPMg9oElAhM9jjrK6wjMflBV/A3N+PwNVPkbqQqsnYd6y77G1mMtD/sqfwNVLQ9+fyO1DU2ufi8tMY60pHjSk+JJTYojLTGeAZnJpCbFk57kPJjTkuJJS4wjtWWd2LI/jpSEwDqwLzUhdA/qULNEYKLHf//rrC0heKa+sZmdtfX4ahvYWeMslbUN+PwN+GobA+vdt9s+9BubO//zOy0xjozkBDKS48lIjicrNZG8vqlkJjsP9fQk51h6cjwZSc7DPKPlWGCdlhgfsQ/tULBEYKLHL3/prK0fwX6ra2yisqaBHTUN7KypZ0dNA5W19YFt53PLg35nbQOVNfXsrG2gpj74X+NpiXFkpiSQleI8rPtnJjOyX3zgoZ7Qus5M/mZfetLun+PsAd7tLBEYE+VUleq6Riqq66nYVc/2XfVUVNdRsaueHYHt7TXO55Z9u4I80BPihN6pifROSaB3agKDe6cwblAmWSkJ9El1HvJZgeNZbZaM5Hji46zFek9kicCYCKSq+GobKa+uo7yqrnW9rbqObYF1xa565/Oueuob22+hkpIQR9+0xNZleG46fVIT6ZuW4DzsUxPok5roPOTTnId7amIcIvZXeTQJWSIQkbnA6UCZqo7v4JypwCwgAdimqseFKh5jIkFTs7Ktuo6tlX7Kquooq/JTXlXnfPY5D/xtVc5Dv75p74d7QpyQnZZETkYi2WlJjOqXQU5GIjlpSc7DPj2R7MBDPzstiZTEOA/u0vQ0oXwjmAfMBh5t76CI9AbuA6ap6tci0i+EsRjjufrGZkp9frb6/JRU+inZWUtJpZ/SwHapz3n4N7VTYdo3LZF+GUnkZiQxIieN3MwkctOd7dyMJPplJJGTnkRWSoL9tW66LGSJQFWXiEh+kFO+Dzyjql8Hzi8LVSwmRsya5enPV9c1UrSjhi07ainaUcuWnbWBzzUUV/rZVl23V5v09KR4BmQlMyAzmZEjcxiQmUz/wHb/TOchn5OeRIKVrZsQ8rKOYDSQICKLgAzgr6ra0dvDTGAmwNChQ8MWoIkwEyeG9PINTc0U76zlq4oaNu+o4evtNRRtr2Xzjho2b69hR03Dbucnxvcir3cKg/ukcOCATAb2TmZQVgoDspIZmJXMgKxkMpITQhqzMW54mQjigcOBE4EU4D0RWaaq6/c8UVXnAHMACgoKutDPz8SUN9901vsxQU1DUzObt9ewsXwXmyp28VVFDZsqdjkP/R21uxXbJMb1YnCfFPL6pDD+4IEM6ZPKkL4pDA48/HPSkqytuokIrhKBiAwAJgEKfKiqW7vht4twKoh3AbtEZAlwCLBXIjDGldtvd9YuEkFlbQMbyqr5oqyaL8qr+aJ8Fxu3VfN1Rc1unZoykuMZlpPGwYOzOGPCIA7ITmVo31SGZqfSPyPZHvQmKnSaCETkR8AtwNuAAH8Tkd+q6tz9/O3ngdkiEg8kAkcAf9nPaxqzm5019azbWsX60io2lFXzeVk1G8qqKauqaz0nMa4XB2SnMqpfOt8eN4DhOWkMz01neE4avVOt8tVEPzdvBDcAh6pqBYCIZAP/BYImAhGZD0wFckSkCLgVp5koqvqAqq4VkVeBT4Bm4CFVXb2vN2JiW11jE411jeyqb+Lv/y7ks8DDv9T3zQM/PSmeEf3SOXZ0LiP7pTOqXzoj+6WT1yfVequamOYmERQBVW22q4DNnX1JVS9wcc6dwJ0uYjCmVWVtA4XFPgpLfKwprqSw2MeGsmoe31IJwCPvfcWofulMGZnDmP4ZjBmQwej+GQzMSra/7o1ph5tEsAV4X0Sex6kjOAv4QER+BqCqd4cwPhPj6hqbWFtSxaqvd/BxUSWrNu/ky227Wo/3y0hi7KBMTjiwH6PeyCA1MY61v51mf+Eb0wVuEsEXgaXF84F1RveHY2Jd8c5aVny1gxVf7eCjzTtZW+xr7UHbLyOJiUN6M+PwPMYPzmLswExyM5K++fKwfzhrSwLGdEmniUBVbwtHICb2NDUrhcU+Pty0nRVf72DlVzsoqfQDzgQfh+T15rKj85mY15uJQ3szILOTop0xY8IUuTHRxU2roVzgRmAckNyyX1VPCGFcJgo1NyuFJT6Wbaxg2cYK3v9yO1X+RgAG906hIL8vhw/tzeEH9OXAgRld70374ovO+owzujlyY6Kbm6KhfwJP4gwgdxVwCVAeyqBM9Niys5ZF68pYsr6cZRu3U1nr9L7Nz07l9AkDOXJ4NpOG9WVgVsr+/9hddzlrSwTGdImbRJCtqg+LyE9VdTGwWEQWhzowE5nqGpv48MsdLFpXxqL15Wwoqwacv/hPGdufySOymTwiu3se/MaYbuEmEbQMoFIiIqcBxUBe6EIykWZnTT1vrS3j9cKt/OfzbdTUN5EY14tJw/py/reGMHVMLiNy063ppjE9lJtEcLuIZAHXAX8DMoH/C2lUpscr3lnL62u28nphKe9/uZ2mZmVAZjLnHDaY48f0Y/KIbFITbd4jYyKBm1ZDLwU+VgLHhzYc05OVVfl58eMSXli1hY+LnM5bI/ulc9Vxwzll7AAm5GXZX/3GRKAOE4GI/A2nA1m7VPWakERkepQqfwOvrSnl+VVbWLphG80K4wZl8vNpB3LKuP6MyE33OsRvPPaY1xEYE5GCvREsD1sUpkdpblbe3bCNJ5dv5s3CUuoamxnSN4Wrp47k7EMHMbJfD+1LOGSI1xEYE5E6TASq+oiIxAF/VNUbwhiT8Uipz89Tyzez4MPNFO2opXdqAucWDOHsQwdx2NA+Pb/Y58knnfV553kbhzERJmgdgao2icjh4QrGhF9Ts7J4fRnzP9jM25+V0dSsHDUimxunHci3x/UnKT6CJje//35nbYnAmC5x06zjIxF5AXgKaB3tS1WfCVlUJuT8DU08vaKIv/9nI19V1JCTnsgVxwzn/G8NIT8nzevwjDFh5CYR9AUqgLZDSihgiSACVdY28Piyr/jH0k1sq67jkLwsbvj+oZwydgCJ8TZBujGxyE3z0cvCEYgJrVKfn4ff/ZIn3v+a6rpGjh2dy1XHDWfy8OyeX/ZvjAkpN4POjQbuB/qr6ngRmQCcqaq3hzw6s98qaxu4b9EG5i3dRENTM6dNGMSVxw5n/OAsr0MzxvQQboqG/o4zXeWDAKr6iYg8AVgi6MH8DU089t5XzH5nAz5/A9+ZOJhrTxrN0OxUr0MLnaef9joCYyKSm0SQqqof7FF80BiieMx+am5Wnlu1hbteX8+WnbUcOzqXm6YdyNhBmV6HFno5OV5HYExEcpMItonICAK9jEVkBlAS0qjMPlnx1XZufm4Na0t8jB+cyZ9mTGDKyBh6OM6b56wvvdTLKIyJOG4SwU+AOcCBIrIF+BK4MKRRmS6p8jfwp1fX8fj7XzEoK4W/nj+RMyYMolesTdloicCYfeImEXylqieJSBrQS1WrQh2Uce+1NVu59fk1lFX5ueyoYVx3ymjSkmzUT2OMe26eGF+KyKs4s5S9HeJ4jEulPj+3Pr+GV9ds5cABGTx40eEcMqS312EZYyKQm0QwBjgDp4joYRF5CVigqu+GNDLToX+tKOI3L66hvrGZG6eN4Ypjhnd9fl9jjAlw06GsFlgILBSRPsBfgcVABA1CEx1q65u45fnVPLWiiEnD+vKn706w4SCMMfvNVWGyiBwHnAdMBz4Ezg1lUGZvG8qq+ck/V7K+rIprThzFT08cRVysVQZ35uWXvY7AmIjkpmfxl8AqnLeCG1R1V/BvmO72/Kot/OKZT0lJiOORyyZx7Ohcr0PqmVKjuLOcMSHk5o3gEFX1hTwSsxd/QxO3vVjI/A++5lv5ffjbBYcxICvZ67B6rvvuc9ZXX+1tHMZEmKA1jCIyHXhJRLaJSLmILBaRU91cWETmikiZiKzu5LxviUhToKOaCdhWXcf3HniP+R98zVXHjWD+FUdaEujMwoXOYozpkmBzFl8BXAncyDfTVhYAfxSRPFWd08m15wGzgUeD/EYc8P+A17oQc9TbsrOWix56n+LKWv5+cQEnj+3vdUjGmCgWrGjo/4CjVXV7m31vB94S3sXpbdwhVV0iIvmd/P7/Av8CvuUi1pjwRXk1Fz30PlV1jTz+wyMoyO/rdUjGmCgXrGhI9kgCAKhqRXf8sIgMBr4DPODi3JkislxElpeXl3fHz/dIq7dUcu4D71Hf1MyCmUdaEjDGhEWwROATkUP23BnY1x3DTMwCfq6qTZ2dqKpzVLVAVQtyc6OzxcyHm7ZzwZxlJMX3YuGVkxk3yOYLMMaER7CioeuAF0TkH8AKnNFHvwVcAvygG367AFgQGN46BzhVRBpV9bluuHZEWbSujKseX8GgrBQe+9ERDO6d4nVIkWnRIq8jMCYidZgIVPVdEZmEM7TEpYAAa4AjVXXr/v6wqg5r+Swi84CXYjEJLFlfzhWPLmdUvwwe/eEkctKTvA7JGBNjgvYjUNVS4JZ9ubCIzAemAjkiUgTcCiQErttpvUAs+Ly0ip/8cyUjctOZP/NIslISvA4psv35z876+uu9jcOYCBOy8YpV9YIunHtpqOLoqSqq67j8kQ9JSojj4Uu/ZUmgO7z0krO2RGBMl9iQlR6oa2ziqsdXUOar4+8XH251AsYYT3UpEYhILxGJgclvQ0dV+cUzn/Lhph3cde4hHDq0j9chGWNiXKeJQESeEJHMwAxlhcA6Ebkh9KFFp/sXf8EzK7fwfyeN5vQJg7wOxxhjXL0RjA0MOnc28DIwFLgolEFFq1dXl/CnV9dx5iGDuObEkV6HE31SUpzFGNMlbiqLE0QkAScRzFbVBhHR0IYVfVZvqeTaJ1dx2NDe/GnGBAL9J0x3euUVryMwJiK5eSN4ENgEpAFLROQAwIal7gJ/QxPXzP+IvqmJPHhRAckJNrmbMabncDNV5T3APW12fSUix4cupOhz7zsb2LhtF4/9cBK5GdZhLGR+9ztn/etfexuHMRHGTWVxfxF5WEReCWyPxRlmwrjw2VYf9y/6gnMOG8wxo6JznKQe4623nMUY0yVuiobm4cwX0NLEZT1wbYjiiSpNzcpN//qUzJQEbj5trNfhGGNMu9wkghxVXQg0A6hqI9DpiKEGHntvE6s27+SW08fSNy3R63CMMaZdbhLBLhHJxhl9FBE5EqgMaVRRoHhnLXe+to5jR+dy1kTrL2CM6bncNB/9GfACMEJElgK5gM0vHISq8uvnVtOscMfZ462paLhkZ3sdgTERyU2roZUichwwBmco6nWq2hDyyCLYS5+U8NZnZdx82kEM6ZvqdTix41//8joCYyJSsMnrz+ng0GgRQVWfCVFMEW1nTT23vbiGCXlZXDZlWOdfMMYYjwV7IzgjyDEFLBG04/cvr2VHTQOPXn4Ecb2sSCisfvELZ/2HP3gbhzERJtgMZZeFM5BosKGsmoXLi5h57HDGDrJBWsPuvfe8jsCYiOSmQ1m2iNwjIitFZIWI/DXQisjs4e9LNpIU34srjx3udSjGGOOam+ajC4By4Ls4rYXKgSdDGVQkKvP5efajLZxbMIRsm3fYGBNB3DQf7auqv2uzfbuInB2ieCLWP/67icbmZn50jFUQG2Mii5s3gndE5PzA7GS9RORc4N+hDiySVPkbeHzZV0wfP5ADstO8Did25eU5izGmS9y8EVyJ06ns8cB2L5zexj8DVFVjvlZ0wQebqfI3MtPqBrz1+OOdn2OM2YubDmUZ4QgkUtU3NvPwu18yeXg2hwzp7XU4xhjTZW7eCBCRCUB+2/OtQ5njhY+L2erz88fvHux1KObaa531rFleRmFMxOk0EYjIXGACsIbACKRYhzIAmpuVOUu+4MABGRw32uYa8NyqVV5HYExEcvNGcKSq2mD67Vi0voz1pdX85bxDbGA5Y0zEctNq6L3ArGRmDw8s3sigrGROn2DDTBtjIpebN4JHcJLBVqAOZwRSVdUJIY2sh1v59Q4++HI7N592EAlxbvKpMcb0TG4SwVzgIuBTvqkjiHlzFm8kMzme8ycN9ToU02L0aK8jMCYiuUkEX6vqC129cKCS+XSgTFXHt3P8QuDngc1q4Meq+nFXf8cLX27bxWuFW7l66gjSk1w1vDLhMGeO1xEYE5HcPMU+E5EngBdxioYAV81H5wGzgUc7OP4lcJyq7hCR6cAc4AgX8XjuxY+LAbjoyHxvAzHGmG7gJhGk4CSAU9rs67T5qKouEZH8IMf/22ZzGRAxYwO8UVjKxCG9GZCV7HUopq2ZM521vRkY0yVuehaHY16CHwKvdHRQRGYCMwGGDvW2TL6kspZPt1Ry47QxnsZh2rF+vdcRGBOR3HQoS8Z5UI8DWv8EVtXLuyMAETk+cP2jOzpHVefgFB1RUFCg3fG7++rNwlIAThnb38swjDGm27hp9/gYMAD4NrAYpwinqjt+PDB0xUPAWapa0R3XDLXXC0sZlpPGiNx0r0Mxxphu4SYRjFTVXwO7VPUR4DRgvwfWEZGhOPUMF6lqRLzT+/wNLNtYwclj+1tPYmNM1HBTWdwQWO8UkfHAVpwB6IISkfnAVCBHRIqAW4EEAFV9ALgFyAbuCzxUG1W1oIvxh9XideU0NCknW7FQzzRxotcRGBOR3CSCOSLSB/g18AKQHvgclKpe0MnxHwE/chNkT/FGYSnZaYkcNrSP16GY9tioo8bsEzethh4KfFwMxOzMKw1Nzbyzroxp4wYQ18uKhYwx0aPDOgIROUNEDmizfYuIfCwiL4hIzE3M+/7G7VT5G61YqCf7wQ+cxRjTJcEqi+8AygFE5HTgB8DlOMVDD4Q+tJ7ljcKtJCf04phRNu9Aj1VU5CzGmC4JlghUVWsCn88BHlbVFYGioph6GqoqbxSWcvTIXFIS47wOxxhjulWwRCAiki4ivYATgbfaHIupsRXWFPsorvRbJzJjTFQKVlk8C1gF+IC1qrocQEQOBUpCHlkP8kZhKSJwwkH9vA7FGGO6XYeJQFXnishrQD+g7fDQW4FwjD/UY7xRWMrhQ/uQk57kdSgmmMmTvY7AmIgUtPmoqm4BtuyxL6beBop21FBY4uMX0w/0OhTTmT/8wesIjIlINsdiJ1oGmbNmo8aYaGWJoBNvrC1lRG4aw22QuZ7vu991FmNMl3QpEYhImoj8QET+HaqAepLK2gbe37idk8cO8DoU40ZFhbMYY7qk00QgIokicraILMRpLXQiMdKhbNG6MhqbbZA5Y0x067CyWEROBi7AmYfgHZx5CSaFacayHmHRunJy0hM5dEhvr0MxxpiQCdZq6DXgP8DRqvolgIj8NSxR9RCrt1QycUgfetkgc8aYKBYsERwOnA+8KSIbgQVAzIyv4G9o4ovyaqYfPNDrUIxbJ57odQTGRKRgHco+Aj4Cfi4iU3CKiRJF5BXg2cA8wlFr3dYqmhXGDsz0OhTj1q87nSbDGNMOV62GVHWpqv4PMBhn6Imo78JZWOIDYNwgSwTGmOgWrLJ4aAeH1uJMOxnVCot9ZCTFk9cnxetQjFvTpzvrV17xNg5jIkywOoJ/Awq0rSlVnCGo+xHl9QWFJT4OGpRpk9RHktparyMwJiJ1WDSkqger6oTA+mDgDGApUA1cG6b4PNHcrKwt8Vn9gDEmJrjpUDZKROYBrwArgLGq+rdQB+alr7bXUFPfZInAGBMTgtURjAd+BYwD/gT8UFWbwhWYlwqLnYrisVZRbIyJAcHqCD4GNuPUFUwCJrUtL1fVa0IbmncKSyqJ7yWM7GcDzUWU00/3OgJjIlKwRHB52KLoYQqLfYzsl05yQlTXh0ef66/3OgJjIlKwDmWPdHRMRA4ITTg9Q2GJjykjcrwOwxhjwiJoZbGITBaRGSLSL7A9QUSeAN4NS3Qe2FZdR6mvzuoHItHUqc5ijOmSDhOBiNwJzAW+C/xbRG4F3gDeB0aFJ7zwWxvoUWwthowxsSJYHcFpwKGq6heRPkAxMEFVPw9PaN5oaTF0kCUCY0yMCFY0VKuqfgBV3QGs60oSEJG5IlImIqs7OC4ico+IbBCRT0TksK6FHhqFJT4GZSXTJy3R61CMMSYsgr0RjBCRF9ps57fdVtUzO7n2PGA28GgHx6fjFDGNAo4A7g+sPVVY7LP6AWNMTAmWCM7aY/uurlxYVZeISH4n139UVRVYJiK9RWSgqpZ05Xe6k7+hiY3bdjF9vM1RHJHOPdfrCIyJSMGajy4GEJF+qlrW9piIjOmG3x6M02GtRVFgn2eJYH1pFU3NavUDkerqq72OwJiI5GY+gv+ISOufWiJyHfBsN/x2e8N6arsniswUkeUisry8vLwbfrp9NrREhKupcRZjTJe4SQRTgYtE5CkRWQKMxhlyYn8VAUPabOfhtEzai6rOUdUCVS3Izc3thp9uX2GJj/SkeIb0SQ3Zb5gQOvVUZzHGdEmniSBQZv8qzqxk+Tjl+tXd8NsvABcHWg8dCVR6WT8AzhvBQQMzbLJ6Y0xMCVZZDICIvIFTbj8e56/2uSKyRFWDDuwiIvNx3iZyRKQIZ1azBABVfQB4GTgV2ADUAJft+23sv5Y5CGYcnudlGMYYE3adJgLgXlV9LvB5p4hMBn7Z2ZdU9YJOjivwExe/HxZfb69hV32T1Q8YY2KOm6Kh50TkaBFp+Yu9D/B4aMMKv8LWoSWyPI7EGGPCy03R0K1AATAG+AeQiJMIpoQ2tPAqLPYR10sY1d/mIIhYl17qdQTGRCQ3RUPfAQ4FVgKoarGIZIQ0Kg8UlvgYmWtzEEQ0SwTG7BM3zUfrA+X5CiAiaaENyRs2tEQU2LbNWYwxXeLmjWChiDwI9BaRK3BmLvt7aMMKr4rqOrb6/Db0dKSbMcNZL1rkaRjGRJpOE4Gq/llETgZ8OPUEt6jqGyGPLIzWllQB1qPYGBOb3LwREHjwR9XDv63CkkrA5iAwxsQmN3UEUa+w2MeAzGT62hwExpgYZIkAp8WQFQsZY2JVsDmL3wqs/1/4wgm/xqZmvijfxZgBUdciNvb8+MfOYozpkmB1BANF5DjgTBFZwB7DRqvqypBGFibbqutpalYG907xOhSzv847z+sIjIlIwRLBLcBNOAPN3b3HMQVOCFVQ4VTq8wPQPzPZ40jMftscmOdoyJDg5xljdhNshrKngadF5Neq+rswxhRWWwOJYIAlgsh30UXO2voRGNMlbvoR/E5EzgSODexapKovhTas8ClrfSNI8jgSY4zxRqethkTkD8BPgcLA8tPAvqiw1ecnrpeQnW6JwBgTm9x0KDsNmKiqzQAi8gjwEfCLUAYWLqW+OnLTk4izWcmMMTHKbT+C3m0+R9WA/aU+P/2zrH7AGBO73LwR/AH4SETewWlCeixR8jYATiLIz47KAVVjz3XXeR2BMRHJTWXxfBFZBHwLJxH8XFW3hjqwcCn11XHEsGyvwzDd4YwzvI7AmIjkdtC5EuCFEMcSdv6GJiprGxhgRUPRYd06Zz1mjLdxGBNhXCWCaNXSmaxfhrUYigpXXumsrR+BMV0S04POba0MdCazNwJjTAwLmghEpJeIrA5XMOFWWlUH2PASxpjYFjQRBPoOfCwiQ8MUT1iV2ThDxhjjqo5gILBGRD4AdrXsVNUzQxZVmGyt9JOc0IvM5JiuKjHGxDg3T8DbQh6FR0qr6uifmYyI9SqOCjff7HUExkQkN/0IFovIAcAoVX1TRFKBuNCHFnqllX4rFoomJ53kdQTGRCQ3g85dATwNPBjYNRh4LoQxhU1plSWCqLJqlbMYY7rETfPRnwBTAB+Aqn4O9AtlUOGgqmyt9DPAhp+OHtde6yzGmC5xkwjqVLW+ZUNE4nFmKOuUiEwTkXUiskFEbmrneJaIvCgiH4vIGhG5zH3o+8dX20hdY7O9ERhjYp6bRLBYRH4JpIjIycBTwIudfUlE4oB7genAWOACERm7x2k/AQpV9RBgKnCXiCR2If59VloV6FVsicAYE+PcJIKbgHLgU+BK4GXATfOMScAGVd0YeKNYAJy1xzkKZIjTbCcd2A40uox9v7T2KrZEYIyJcW5aDTUHJqN5H+fBvU5V3RQNDQY2t9kuAo7Y45zZOIPZFQMZwHktE+C0JSIzgZkAQ4d2T9+2Upui0hhjABeJQEROAx4AvsAZhnqYiFypqq909tV29u2ZQL4NrAJOAEYAb4jIf1TVt9uXVOcAcwAKCgpc1U90ptR6FUef3//e6wiMiUhuOpTdBRyvqhsARGQE8G+gs0RQBAxps52H85d/W5cBfwy8YWwQkS+BA4EPXMS1X0p9dWSlJJCcEBVdIgzAUUd5HYExEclNHUFZSxII2AiUufjeh8AoERkWqAA+n73nNPgaOBFARPoDYwLXD7lSn9/qB6LNf//rLMaYLunwjUBEzgl8XCMiLwMLcYp2vofzkA9KVRtF5H+A13B6Is9V1TUiclXg+APA74B5IvIp38x+tm1/bsitUp+fflY/EF1++UtnbfMRGNMlwYqG2s77VwocF/hcDvRxc3FVfRmnlVHbfQ+0+VwMnOIq0m5W6qtjVP8ML37aGGN6lA4TgaqGrXNXuDU1K+XVdVY0ZIwxuGs1NAz4XyC/7fmRPAx1RXUdTc1qTUeNMQZ3rYaeAx7G6U28Vxv/SLTVmo4aY0wrN4nAr6r3hDySMCr12RSVUWnWLK8jMCYiuUkEfxWRW4HXgbqWnaq6MmRRhVhLZzKbtD7KTJzodQTGRCQ3ieBg4CKc3r8tRUMa2I5IpT4/vQSy08Iyvp0JlzffdNY2QU1EaGhooKioCL/f73UoUSU5OZm8vDwSEhJcf8dNIvgOMLztUNSRrtTnJyc9ifg4N/3pTMS4/XZnbYkgIhQVFZGRkUF+fr5NF9tNVJWKigqKiooYNmyY6++5eRJ+DPTe18B6oq2+OisWMsZjfr+f7OxsSwLdSETIzs7u8luWmzeC/sBnIvIhu9cRRGzz0TKfn7w+qV6HYUzMsyTQ/fblv6mbRHBr10Pp2Up9fgryXXWONsaYqOdmPoLF4QgkXPwNTeyoaaB/hhUNGWNCa9asWcycOZPU1J5dAtFpHYGIVImIL7D4RaRJRHydfa+nKq8K9CGwOoLo8+CDzmJMDzFr1ixqamq8DqNTbt4IdhuZTUTOxpmGMiJZr+IoNmaM1xGYfXTbi2soLO7evy/HDsrk1jPGdXreo48+yp///GdEhAkTJnD77bdz+eWXU15eTm5uLv/4xz8YOnQol156KaeffjozZswAID09nerqahYtWsRvfvMbcnJyWL16NYcffjiPP/44f/vb3yguLub4448nJyeHd955p1vvrzu5qSPYjao+JyI3hSKYcLApKqPYiy866zPOCH6eMQFr1qzhjjvuYOnSpeTk5LB9+3YuueQSLr74Yi655BLmzp3LNddcw3PPPRf0Oh999BFr1qxh0KBBTJkyhaVLl3LNNddw9913884775CTkxOeG9pHbgadO6fNZi+ggL2nnIwYNml9FLvrLmdtiSDiuPnLPRTefvttZsyY0fqg7tu3L++99x7PPPMMABdddBE33nhjp9eZNGkSeXl5AEycOJFNmzZx9NFHhy7wbubmjaDtv6pGYBNwVkiiCYOyqjoS43uRleK+150xJjqpaqfNLVuOx8fH09zc3Pq9+vpv+tgmJX1TwhAXF0djY2MIog2dTiuLVfWyNssVqnqHqrqZqrJHapmi0tovG2NOPPFEFi5cSEVFBQDbt2/nqKOOYsGCBQD885//bP3LPj8/nxUrVgDw/PPP09DQ0On1MzIyqKqqClH03SfYVJW3BPmequrvQhBPyG2t9Fv9gDEGgHHjxvGrX/2K4447jri4OA499FDuueceLr/8cu68887WymKAK664grPOOotJkyZx4oknkpaW1un1Z86cyfTp0xk4cGCPriwW1faL+0XkunZ2pwE/BLJVNT2UgXWkoKBAly9fvs/fP/7Pixg3KJPZ3z+sG6MyPcLUqc7a5iyOCGvXruWggw7yOoyo1N5/WxFZoaoF7Z0fbKrKu9pcIAP4KXAZsAC4q6Pv9WSqytZKPycc2M/rUEwoPPaY1xEYE5GCVhaLSF/gZ8CFwCPAYaq6IxyBhUJVXSO1DU1WNBSthgzxOgJjIlKwOoI7gXOAOcDBqlodtqhCpMw6k0W3J5901ued520cxkSYYG8E1+GMNnoz8Ks2rWwEp7I4M8SxdbutlTZFZVS7/35nbYnAmC4JVkcQdbO2tE5RaYnAGGNaRd3DPhgbZ8gYY/YWU4mgzOcnMzmelMQ4r0Mxxnhs586d3Hfffa3bixYt4vTTT+/237n00kt5+umnu/SdCy+8kDFjxjB+/Hguv/zy1s5rqso111zDyJEjmTBhAitXruyWGGMqEWz1+e1twBgD7J0I3GpqagpBNLu78MIL+eyzz/j000+pra3loYceAuCVV17h888/5/PPP2fOnDn8+Mc/7pbfi6lEUOqrs0QQzZ5+2llMZJo6de+l5UFdU9P+8XnznOPbtu19rBM33XQTX3zxBRMnTuSGG24AoLq6mhkzZnDggQdy4YUX0tLhNj8/n9/+9rccffTRPPXUU7z++utMnjyZww47jO9973tUV1e3XnPs2LFMmDCB66+/vvW3lixZwlFHHcXw4cNdvR2ceuqpiAgiwqRJkygqKgKcoS0uvvhiRIQjjzySnTt3UlJS0un1OhPSRCAi00RknYhs6GjoahGZKiKrRGSNiIR0NrQyeyOIbjk5zmKMC3/84x8ZMWIEq1at4s477wSc4aRnzZpFYWEhGzduZOnSpa3nJycn8+6773LSSSdx++238+abb7Jy5UoKCgq4++672b59O88++yxr1qzhk08+4eabb279bklJCe+++y4vvfQSN930zaNw4sSJQWNsaGjgscceY9q0aQBs2bKFIW36y+Tl5bFly5b9/m/R5fkI3BKROOBe4GSgCPhQRF5Q1cI25/QG7gOmqerXIhKyLr/NzUpZVZ11JotmLX8dXnqpl1GYfRVsaJDU1ODHc3K6ZWiRYMNJnxdolrxs2TIKCwuZMmUKAPX19UyePJnMzEySk5P50Y9+xGmnnbZbfcPZZ59Nr169GDt2LKWlpa37V61aFTSeq6++mmOPPZZjjjkGgPaGBOqOATRDlghwZjHboKobAURkAc7w1YVtzvk+8Iyqfg0QylFNK3bV09isDLApKqOXJQKzn4INJ90yyJyqcvLJJzN//vy9vv/BBx/w1ltvsWDBAmbPns3bb7+913U7Gt9tT7fddhvl5eU82Gb61by8PDZv3ty6XVRUxKBBg1zeXcdCWTQ0GNjcZrsosK+t0UAfEVkkIitE5OL2LiQiM0VkuYgsLy8v36dgWvoQ9LNJ640x7PsQ0UceeSRLly5lw4YNANTU1LB+/Xqqq6uprKzk1FNPZdasWZ3+tR/MQw89xGuvvcb8+fPp1eubx/SZZ57Jo48+iqqybNkysrKyGDhw4D7/TotQvhG0976yZyqMBw4HTgRSgPdEZJmqrt/tS6pzcIa6oKCgYJ9mR2vtTGZvBMYYIDs7mylTpjB+/HimT5/Oaaed5up7ubm5zJs3jwsuuIC6Ome0gttvv52MjAzOOuss/H4/qspf/vKXTq81ceLEdhPGVVddxQEHHMDkyZMBOOecc7jllls49dRTefnllxk5ciSpqamtQ2Tvr1AmgiKg7ShgeUBxO+dsU9VdwC4RWQIcAqynm2WlJDBt3AAG9bZEYIxxPPHEE7ttT23T2mj27Nmtnzdt2rTbeSeccAIffvjhXtf74IMP9to3r6XIMqClhRF0XEfQ0QxnIsK9997b7rH9EcpE8CEwSkSGAVuA83HqBNp6HpgtIvFAInAE0Hka3QcF+X0pyO8biksbY0xEC1kiUNVGEfkf4DUgDpirqmtE5KrA8QdUda2IvAp8AjQDD6nq6lDFZKLcyy97HYExESmUbwSo6svAy3vse2CP7TuBO0MZh4kRqaleR2C6yM3k8aZr3LZKaiumehabKHfffd/0RDU9XnJyMhUVFfv04DLtU1UqKipITu5aXWhI3wiMCauFC5311Vd7G4dxJS8vj6KiIva1SbhpX3JycmunOLcsERhjPJGQkMCwYcO8DsNgRUPGGBPzLBEYY0yMs0RgjDExTiKtxl5EyoGvOjktB9gWhnB6oli+d4jt+4/le4fYvn83936Aqua2dyDiEoEbIrJcVQu8jsMLsXzvENv3H8v3DrF9//t771Y0ZIwxMc4SgTHGxLhoTQRzvA7AQ7F87xDb9x/L9w6xff/7de9RWUdgjDHGvWh9IzDGGOOSJQJjjIlxUZcIRGSaiKwTkQ0icpPX8YSSiMwVkTIRWd1mX18ReUNEPg+s+3gZY6iIyBAReUdE1orIGhH5aWB/rNx/soh8ICIfB+7/tsD+mLh/ABGJE5GPROSlwHYs3fsmEflURFaJyPLAvn2+/6hKBCISB9wLTAfGAheIyFhvowqpecC0PfbdBLylqqOAtwLb0agRuE5VDwKOBH4S+H8dK/dfB5ygqocAE4FpInIksXP/AD8F1rbZjqV7BzheVSe26T+wz/cfVYkAmARsUNWNqloPLADO8jimkFHVJcD2PXafBTwS+PwIcHY4YwoXVS1R1ZWBz1U4D4TBxM79q6q2TH6bEFiUGLl/EckDTgMearM7Ju49iH2+/2hLBIOBzW22iwL7Ykl/VS0B52EJ9PM4npATkXzgUOB9Yuj+A0Ujq4Ay4A1VjaX7nwXciDPFbYtYuXdwkv7rIrJCRGYG9u3z/UfbfATtzXln7WOjmIikA/8CrlVVXyxNe6iqTcBEEekNPCsi4z0OKSxE5HSgTFVXiMhUj8PxyhRVLRaRfsAbIvLZ/lws2t4IioAhbbbzgGKPYvFKqYgMBAisyzyOJ2REJAEnCfxTVZ8J7I6Z+2+hqjuBRTj1RbFw/1OAM0VkE07x7wki8jixce8AqGpxYF0GPItTLL7P9x9tieBDYJSIDBORROB84AWPYwq3F4BLAp8vAZ73MJaQEedP/4eBtap6d5tDsXL/uYE3AUQkBTgJ+IwYuH9V/YWq5qlqPs6/8bdV9QfEwL0DiEiaiGS0fAZOAVazH/cfdT2LReRUnPLDOGCuqt7hbUShIyLzgak4Q9CWArcCzwELgaHA18D3VHXPCuWIJyJHA/8BPuWbcuJf4tQTxML9T8CpEIzD+YNuoar+VkSyiYH7bxEoGrpeVU+PlXsXkeE4bwHgFO8/oap37M/9R10iMMYY0zXRVjRkjDGmiywRGGNMjLNEYIwxMc4SgTHGxDhLBMYYE+MsEZiYISLZgdEaV4nIVhHZEvi8U0QKQ/B7vxGR67v4neoO9s8TkRndE5kxu7NEYGKGqlYERmucCDwA/CXweSK7j1nTLhGJtiFZjAEsERjTIk5E/h4Y2//1QG9dRGSRiPxeRBYDPxWRw0VkcWCwr9fadOm/RkQKReQTEVnQ5rpjA9fYKCLXtOwUkZ+JyOrAcu2ewYhjduCa/ya6B1AzHrO/cIxxjAIuUNUrRGQh8F3g8cCx3qp6XGBso8XAWapaLiLnAXcAl+OM/T5MVetahn4IOBA4HsgA1onI/cAE4DLgCJyBEt8XkcWq+lGb730HGAMcDPQHCoG5obhxYywRGOP4UlVXBT6vAPLbHHsysB4DjMcZ7RGc4R1KAsc+Af4pIs/hDPPR4t+qWgfUiUgZzkP9aOBZVd0FICLPAMcAbRPBscD8wAijxSLy9v7fojHts0RgjKOuzecmIKXN9q7AWoA1qjq5ne+fhvPwPhP4tYiM6+C68bQ/XHp7bPwXExZWR2CMe+uAXBGZDM4w2CIyTkR6AUNU9R2cyVJ6A+lBrrMEOFtEUgOjR34HZwC9Pc85PzD5zECc4iVjQsLeCIxxSVXrA0047xGRLJx/P7OA9cDjgX2C0xppZ0eT5KjqShGZB3wQ2PXQHvUD4IwueQLO6KrrceomjAkJG33UGGNinBUNGWNMjLNEYIwxMc4SgTHGxDhLBMYYE+MsERhjTIyzRGCMMTHOEoExxsS4/w9GPgm6lfgfnwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot('threshold', 'count', data=exRNA_filter_results)\n",
    "plt.xlabel(\"Threshold\")\n",
    "plt.ylabel(\"Number of exRNA Samples Overlap\")\n",
    "ax = plt.gca()\n",
    "cutoff=20\n",
    "ax.axvline(cutoff, color = \"r\", ls=\"--\", label=f\"thresh: {cutoff}\")\n",
    "plt.legend()\n",
    "plt.savefig(\"figures/exRNA_atlas_samples_filter_thresh.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we choose a cutoff of 20. If a locus is found (overlapped) in at least 20 exRNA samples, we will remove from consideration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "165016"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_remove = set(exRNA_loci_hits_df[exRNA_loci_hits_df[\"counts\"] >= 20][\"loci\"])\n",
    "len(to_remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1569283"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exRNA_filtered_loci = []\n",
    "for locus in loci_map:\n",
    "    if locus not in to_remove:\n",
    "        exRNA_filtered_loci.append(locus)\n",
    "len(exRNA_filtered_loci)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this exRNA cleanup step, we removed about 9.5% of our feature loci. Removing 165016 loci from 1734299."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed\", \"w\") as out: \n",
    "    for locus in exRNA_filtered_loci:\n",
    "        splits = locus.split(\":\")\n",
    "        start = splits[1].split(\"-\")[0]\n",
    "        end = splits[1].split(\"-\")[1]\n",
    "        bed = f\"{splits[0]}\\t{start}\\t{end}\\t{locus}\\t.\\t{splits[-1]}\"\n",
    "        out.write(bed + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Counts\n",
    "Last preprocess step for the cell lines is to create counts for each cell line sample for each smRNA loci feature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for f in data/IC/*.filter.bed;\n",
    "do\n",
    "out=${f/.filter.bed/.exRNA.intersect.bed}\n",
    "echo \"intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed > $out\"\n",
    "intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map_exRNA_filtered.bed > $out\n",
    "done &> log/cell_lines_exRNA_intersect.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for f in data/IC/*.filter.bed;\n",
    "do\n",
    "out=${f/.filter.bed/.intersect.bed}\n",
    "echo \"intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map.bed > $out\"\n",
    "intersectBed -s -wo -a $f -b data/cell_lines_smRNAs_loci_map.bed > $out\n",
    "done &> log/cell_lines_intersect.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_loci = {}\n",
    "cpm_map = {}\n",
    "exRNA_filtered_cell_lines = [f for f in os.scandir(\"data/IC/\") if f.name.endswith(\".exRNA.intersect.bed\")]\n",
    "for f in exRNA_filtered_cell_lines:  \n",
    "    \n",
    "    sample = f.name.split(\".\")[0]\n",
    "    bed_file = pd.read_csv(f\"data/IC/{sample}.filter.bed\", header=None, sep=\"\\t\")\n",
    "    assert len(bed_file[3].unique()) == bed_file.shape[0] #Ensures number of unique reads (query_id) matches total number of reads in bed file, so that we are not over counting\n",
    "    #cpm calculations: total reads within pre-exRNA filtered bedfile. Note reads that had \"chrUn\" or low-complexity seqeunces where filered out and not included in cpm normalization.           \n",
    "    cpm = 1000000/bed_file.shape[0]\n",
    "    cpm_map[sample] = cpm\n",
    "    \n",
    "    #Counts of exRNA filtered loci\n",
    "    ex_fil_bed = pd.read_csv(f, header=None, sep=\"\\t\")\n",
    "    loci_features = ex_fil_bed[9] #Locus annotation as a feature.\n",
    "    loci_bp_overlaps = ex_fil_bed[12] #Number of bp overlaps between read and locus annotation.\n",
    "    read_ids = ex_fil_bed[3] #Query ID from original bamfile for each read.\n",
    "    \n",
    "    #First create read_id map\n",
    "    read_id_locus_map = {}\n",
    "    for i in range(len(read_ids)):\n",
    "        _id = read_ids.iloc[i]\n",
    "        num_bp_match = loci_bp_overlaps.iloc[i]\n",
    "        locus = loci_features.iloc[i]\n",
    "        \n",
    "        if _id in read_id_locus_map: #Indicates multiple hits/overlaps for one read.\n",
    "            if num_bp_match > read_id_locus_map[_id][\"bp\"]: #Update locus count if more number of bp matched. This is our simple binning procedure.\n",
    "                read_id_locus_map[_id] = {\"locus\":locus, \"bp\":num_bp_match}\n",
    "        else:\n",
    "            read_id_locus_map[_id] = {\"locus\":locus, \"bp\":num_bp_match}\n",
    "            \n",
    "    assert len(read_id_locus_map) == len(ex_fil_bed[3].unique()) #Ensures we do not overcount reads   \n",
    "    \n",
    "    #Counts of loci\n",
    "    sample_loci[sample] = {}\n",
    "    for _id in read_id_locus_map: \n",
    "        locus = read_id_locus_map[_id][\"locus\"] \n",
    "        if locus in sample_loci[sample]:\n",
    "            sample_loci[sample][locus] += 1\n",
    "        else:\n",
    "            sample_loci[sample][locus] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/counts/cell_lines_loci_counts.json', 'w') as f:\n",
    "    js.dump(sample_loci, f)\n",
    "    f.close()\n",
    "    \n",
    "with open (\"data/counts/cell_lines_cpm_map.json\", \"w\") as f:\n",
    "    js.dump(cpm_map, f)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess TCGA Data\n",
    "Since TCGA datasets are quite bit larger, we run the following pre-processing steps as scripts instead of in this notebook. All scripts can be found in the `scripts` directory. <br>\n",
    "TCGA dataset we use in this project involves: 679 normal samples from varying tissues and 1103 BRCA cancer samples.\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dust \n",
    "Eliminate low-complexity sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`python3 scripts/dust_filter.py &> log/dust_filter_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bamtobed\n",
    "Convert all the filtered bam files to bed files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`ls /rumi/shams/jwang/BRCA_oncRNA/data/TCGA/*.dust.bam | parallel -j 30 bash scripts/bamtobed.sh {} &> log/bamtobed_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter TCGA Data\n",
    "TCGA bed data contains certain formats that are not compatible with bedtools and are not of interest (ex: chrUn). Here we filter our TCGA data before further processing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command used in smRNA env: <br>\n",
    "`python3 scripts/filter_TCGA.py &> log/filter_TCGA.out`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "Finished preprocessing and creating counts for loci found in our cell lines and preprocessing TCGA data. <br>\n",
    "Next step is identify cancer-enriched RNAs."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:smRNA]",
   "language": "python",
   "name": "conda-env-smRNA-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
